



<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
      
      
        <meta name="author" content="Marcus D. R. Klarqvist">
      
      
        <meta name="lang:clipboard.copy" content="Copy to clipboard">
      
        <meta name="lang:clipboard.copied" content="Copied to clipboard">
      
        <meta name="lang:search.language" content="en">
      
        <meta name="lang:search.pipeline.stopwords" content="True">
      
        <meta name="lang:search.pipeline.trimmer" content="True">
      
        <meta name="lang:search.result.none" content="No matching documents">
      
        <meta name="lang:search.result.one" content="1 matching document">
      
        <meta name="lang:search.result.other" content="# matching documents">
      
        <meta name="lang:search.tokenizer" content="[\s\-]+">
      
      <link rel="shortcut icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.0.4, mkdocs-material-3.2.0">
    
    
      
        <title>R - Tomahawk</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/application.572ca0f0.css">
      
      
    
    
      <script src="../assets/javascripts/modernizr.8c900955.js"></script>
    
    
      
        <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700|Roboto+Mono">
        <style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
      
    
    <link rel="stylesheet" href="../assets/fonts/material-icons.css">
    
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
    
  </head>
  
    <body dir="ltr">
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448"
    viewBox="0 0 416 448" id="__github">
  <path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19-18.125
        8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19 18.125-8.5
        18.125 8.5 10.75 19 3.125 20.5zM320 304q0 10-3.125 20.5t-10.75
        19-18.125 8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19
        18.125-8.5 18.125 8.5 10.75 19 3.125 20.5zM360
        304q0-30-17.25-51t-46.75-21q-10.25 0-48.75 5.25-17.75 2.75-39.25
        2.75t-39.25-2.75q-38-5.25-48.75-5.25-29.5 0-46.75 21t-17.25 51q0 22 8
        38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0
        37.25-1.75t35-7.375 30.5-15 20.25-25.75 8-38.375zM416 260q0 51.75-15.25
        82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5-41.75
        1.125q-19.5 0-35.5-0.75t-36.875-3.125-38.125-7.5-34.25-12.875-30.25-20.25-21.5-28.75q-15.5-30.75-15.5-82.75
        0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25
        30.875q36.75-8.75 77.25-8.75 37 0 70 8 26.25-20.5
        46.75-30.25t47.25-9.75q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34
        99.5z" />
</svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
    
      <a href="#getting-started-with-rtomahawk" tabindex="1" class="md-skip">
        Skip to content
      </a>
    
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href=".." title="Tomahawk" class="md-header-nav__button md-logo">
          
            <img src="../images/twk_logo_white.svg" height="24">
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          
            
              <span class="md-header-nav__topic">
                Tomahawk
              </span>
              <span class="md-header-nav__topic">
                R
              </span>
            
          
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          
            <label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
            
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="__search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
          
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  


  <a href="https://github.com/mklarqvist/tomahawk/" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#__github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
        
      
      
      <main class="md-main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <a href=".." title="Tomahawk" class="md-nav__button md-logo">
      
        <img src="../images/twk_logo_white.svg" width="48" height="48">
      
    </a>
    Tomahawk
  </label>
  
    <div class="md-nav__source">
      


  


  <a href="https://github.com/mklarqvist/tomahawk/" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#__github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href=".." title="Home" class="md-nav__link">
      Home
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../introduction/" title="Introduction" class="md-nav__link">
      Introduction
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../installation/" title="Installation" class="md-nav__link">
      Installation
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-4" type="checkbox" id="nav-4">
    
    <label class="md-nav__link" for="nav-4">
      Technical details
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-4">
        Technical details
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../algorithm-overview/" title="Algorithm overview" class="md-nav__link">
      Algorithm overview
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../aggregation/" title="Aggregation" class="md-nav__link">
      Aggregation
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../job-balancing/" title="Job balancing" class="md-nav__link">
      Job balancing
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-5" type="checkbox" id="nav-5" checked>
    
    <label class="md-nav__link" for="nav-5">
      Tutorial
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-5">
        Tutorial
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../tutorial/" title="CLI" class="md-nav__link">
      CLI
    </a>
  </li>

        
          
          
          

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    
      
    
    
      <label class="md-nav__link md-nav__link--active" for="__toc">
        R
      </label>
    
    <a href="./" title="R" class="md-nav__link md-nav__link--active">
      R
    </a>
    
      
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#about" title="About" class="md-nav__link">
    About
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#prerequisites" title="Prerequisites" class="md-nav__link">
    Prerequisites
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#import-files-into-tomahawk" title="Import files into Tomahawk" class="md-nav__link">
    Import files into Tomahawk
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#loading-and-reading-two-data" title="Loading and reading .two data" class="md-nav__link">
    Loading and reading .two data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#calculating-linkage-disequilibrium" title="Calculating linkage-disequilibrium" class="md-nav__link">
    Calculating linkage-disequilibrium
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#plotting-data-with-rtomahawk" title="Plotting data with rtomahawk" class="md-nav__link">
    Plotting data with rtomahawk
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#color-schemes" title="Color schemes" class="md-nav__link">
    Color schemes
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#square-representation" title="Square representation" class="md-nav__link">
    Square representation
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#triangular-representation" title="Triangular representation" class="md-nav__link">
    Triangular representation
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#visualizing-gwas-data-and-ld" title="Visualizing GWAS data and LD" class="md-nav__link">
    Visualizing GWAS data and LD
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#aggregating-and-visualizing-datasets" title="Aggregating and visualizing datasets" class="md-nav__link">
    Aggregating and visualizing datasets
  </a>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
    
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../docker-tutorial/" title="Docker" class="md-nav__link">
      Docker
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-6" type="checkbox" id="nav-6">
    
    <label class="md-nav__link" for="nav-6">
      Command line interface
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-6">
        Command line interface
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../cli/cli-aggregate/" title="aggregate" class="md-nav__link">
      aggregate
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../cli/cli-calc/" title="calc" class="md-nav__link">
      calc
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../cli/cli-concat/" title="concat" class="md-nav__link">
      concat
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../cli/cli-import/" title="import" class="md-nav__link">
      import
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../cli/cli-sort/" title="sort" class="md-nav__link">
      sort
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../cli/cli-view/" title="view" class="md-nav__link">
      view
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../cli/cli-haplotype/" title="haplotype" class="md-nav__link">
      haplotype
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../cli/cli-decay/" title="decay" class="md-nav__link">
      decay
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../api-documentation/" title="API documentation" class="md-nav__link">
      API documentation
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../developer-documentation/" title="Developer documentation" class="md-nav__link">
      Developer documentation
    </a>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#about" title="About" class="md-nav__link">
    About
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#prerequisites" title="Prerequisites" class="md-nav__link">
    Prerequisites
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#import-files-into-tomahawk" title="Import files into Tomahawk" class="md-nav__link">
    Import files into Tomahawk
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#loading-and-reading-two-data" title="Loading and reading .two data" class="md-nav__link">
    Loading and reading .two data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#calculating-linkage-disequilibrium" title="Calculating linkage-disequilibrium" class="md-nav__link">
    Calculating linkage-disequilibrium
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#plotting-data-with-rtomahawk" title="Plotting data with rtomahawk" class="md-nav__link">
    Plotting data with rtomahawk
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#color-schemes" title="Color schemes" class="md-nav__link">
    Color schemes
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#square-representation" title="Square representation" class="md-nav__link">
    Square representation
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#triangular-representation" title="Triangular representation" class="md-nav__link">
    Triangular representation
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#visualizing-gwas-data-and-ld" title="Visualizing GWAS data and LD" class="md-nav__link">
    Visualizing GWAS data and LD
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#aggregating-and-visualizing-datasets" title="Aggregating and visualizing datasets" class="md-nav__link">
    Aggregating and visualizing datasets
  </a>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/mklarqvist/tomahawk/edit/master/docs/r-tutorial.md" title="Edit this page" class="md-icon md-content__icon">&#xE3C9;</a>
                
                
                <h1 id="getting-started-with-rtomahawk">Getting started with <code>rtomahawk</code><a class="headerlink" href="#getting-started-with-rtomahawk" title="Permanent link">&para;</a></h1>
<h2 id="about">About<a class="headerlink" href="#about" title="Permanent link">&para;</a></h2>
<p>This is an introductory tutorial for using the R-bindings for Tomahawk
(<code>rtomahawk</code>). It will cover:</p>
<ul>
<li>Importing data into Tomahawk</li>
<li>Loading, subsetting, and filtering output data</li>
<li>Computing linkage-disequilibrium</li>
<li>Plotting raw linkage-disequilibrium data</li>
<li>Aggregating and visualizing datasets</li>
<li>Combining LD information with GWAS P-values and annotations</li>
</ul>
<h2 id="prerequisites">Prerequisites<a class="headerlink" href="#prerequisites" title="Permanent link">&para;</a></h2>
<p>In order to follow this tutorial exactly, you will need to download the variant
call data for chromosome 6 the 1000 Genomes Project (1KGP3) cohort and validate
its correctness. You can use the following commands on Debian-based systems, or,
go to the <a href="ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/">1KGP3 FTP
server</a> and download
the data manually.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr6.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.<span class="o">{</span>gz,gz.tbi<span class="o">}</span>
md5sum ALL.chr6.*.gz*
</pre></div>
</td></tr></table>

<table>
<thead>
<tr>
<th>MD5 checksum</th>
<th>File</th>
</tr>
</thead>
<tbody>
<tr>
<td>9be06b094cc80281c9aec651d657abb9</td>
<td>ALL.chr6.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz</td>
</tr>
<tr>
<td>5ed21968b2e212fa2cc9237170866c5b</td>
<td>ALL.chr6.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz.tbi</td>
</tr>
</tbody>
</table>
<p>This tutorial was designed and tested for</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="o">&gt;</span> <span class="nf">tomahawkVersion</span><span class="p">()</span>
<span class="n">rtomahawk</span><span class="o">:</span> <span class="m">0.1.0</span>
<span class="n">Libraries</span><span class="o">:</span> <span class="n">tomahawk</span><span class="m">-0.7.0</span><span class="p">;</span> <span class="n">ZSTD</span><span class="m">-1.3.1</span><span class="p">;</span> <span class="n">htslib</span> <span class="m">1.9</span>
</pre></div>
</td></tr></table>

<h2 id="import-files-into-tomahawk">Import files into Tomahawk<a class="headerlink" href="#import-files-into-tomahawk" title="Permanent link">&para;</a></h2>
<p>Depending on the downstream application you want to import either Tomahawk
representations of sequence variant files (<code>.twk</code>) or Tomahawk-generated output
LD data (<code>.two</code>). Both operations require minimal work. First we will import a
<code>vcf</code>/<code>vcf.gz</code>/<code>bcf</code> file into the binary Tomahawk file format (<code>.twk</code>):</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="n">twk</span> <span class="o">&lt;-</span> <span class="nf">import</span><span class="p">(</span><span class="s">&quot;ALL.chr6.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz&quot;</span><span class="p">,</span><span class="s">&quot;1kgp3_chr6&quot;</span><span class="p">)</span>
</pre></div>
</td></tr></table>

<div class="admonition note">
<p class="admonition-title">Auto-completion of file extensions</p>
<p>You do not have to append the <code>.twk</code> suffix to the output name as Tomahawk
will automatically add this if missing. In the example above <code>"1kgp3_chr6"</code>
will be converted to <code>"1kgp3_chr6.two"</code> automatically. This is true for 
most Tomahawk commands when using the CLI but <em>not</em> when using <code>rtomahawk</code>.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Unsafe method</p>
<p>In the current release of <code>rtomahawk</code>, this subroutine does <em>not</em> check for
user-interruption (for example <code>Ctrl+C</code> or <code>Ctrl+Z</code>) commands. This means that
you need to wait until the underlying process has finished or you have to
terminate the host <code>R</code> session, in turn killing the spawned process. This will
be fixed in upcoming releases.</p>
</div>
<p>By default, <code>rtomahawk</code> will print verbose output to the console during the
importing procedure. This will take several minutes depending on your system.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>[2019-01-21 10:42:59,178][LOG][READER] Opening ALL.chr6.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz...
[2019-01-21 10:42:59,188][LOG][VCF] Constructing lookup table for 86 contigs...
[2019-01-21 10:42:59,188][LOG][VCF] Samples: 2,504...
[2019-01-21 10:42:59,188][LOG][WRITER] Opening 1kgp3_chr6.twk...
[2019-01-21 10:44:41,225][LOG] Duplicate site dropped: 6:18233985
[2019-01-21 10:48:06,492][LOG] Duplicate site dropped: 6:55137646
[2019-01-21 10:49:01,472][LOG] Duplicate site dropped: 6:67839893
[2019-01-21 10:49:36,487][LOG] Duplicate site dropped: 6:74373442
[2019-01-21 10:49:55,386][LOG] Duplicate site dropped: 6:77843171
[2019-01-21 10:53:47,992][LOG] Duplicate site dropped: 6:121316830
[2019-01-21 10:56:08,853][LOG] Duplicate site dropped: 6:148573620
[2019-01-21 10:56:25,301][LOG] Duplicate site dropped: 6:151397786
[2019-01-21 10:58:16,584][LOG] Wrote: 4,784,608 variants to 9,570 blocks...
[2019-01-21 10:58:16,584][LOG] Finished: 15m17,405s
[2019-01-21 10:58:16,584][LOG] Filtered out 239,511 sites (4.76722%):
[2019-01-21 10:58:16,584][LOG]    Invariant: 15,485 (0.308213%)
[2019-01-21 10:58:16,584][LOG]    Missing threshold: 0 (0%)
[2019-01-21 10:58:16,584][LOG]    Insufficient samples: 0 (0%)
[2019-01-21 10:58:16,584][LOG]    Mixed ploidy: 0 (0%)
[2019-01-21 10:58:16,584][LOG]    No genotypes: 0 (0%)
[2019-01-21 10:58:16,584][LOG]    No FORMAT: 0 (0%)
[2019-01-21 10:58:16,584][LOG]    Not biallelic: 26,277 (0.523017%)
[2019-01-21 10:58:16,584][LOG]    Not SNP: 194,566 (3.87264%)
</pre></div>
</td></tr></table>

<p>The <code>import</code> procedure will return a new empty <code>twk</code> class with a file pointer
set to the newly created output file:</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="o">&gt;</span> <span class="n">twk</span>
<span class="n">An</span> <span class="n">object</span> <span class="n">of</span> <span class="n">class</span> <span class="n">twk</span>
</pre></div>
</td></tr></table>

<div class="admonition success">
<p class="admonition-title">Importing variant call archives</p>
<p>Success: This object is now ready to be used by other downstream functions in <code>rtomahawk</code>.</p>
</div>
<h2 id="loading-and-reading-two-data">Loading and reading <code>.two</code> data<a class="headerlink" href="#loading-and-reading-two-data" title="Permanent link">&para;</a></h2>
<p>Many functions in <code>rtomahawk</code> computes or use linkage-disequilibrium and require
a different loading procedure involving the <code>openTomahawkOuput</code> subroutine. In
this example we have a pre-computed <code>.two</code> output file available locally. We
open this file and load all supportive information such as headers, footers,
sample information, and validate the archive and collate this information in a
new <code>twk</code> class together with a file pointer to the target archive:</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="n">twk</span><span class="o">&lt;-</span><span class="nf">openTomahawkOutput</span><span class="p">(</span><span class="s">&quot;example.two&quot;</span><span class="p">)</span>
</pre></div>
</td></tr></table>

<div class="admonition note">
<p class="admonition-title">Auto-completion of file extensions</p>
<p>This function will <em>not</em> search for files in the directly by auto-completing
missing file extensions. This is done purposely to avoid enforcing file
extensions and any possible capitalization issues.</p>
</div>
<p>This procedure is extremely fast as very little data is loaded from the archive
and instead maintaining a data pointer. This approach, albeit a bit akward,
enables us to handle more data than there is available system memory (RAM).</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="o">&gt;</span> <span class="nf">system.time</span><span class="p">(</span><span class="n">twk</span><span class="o">&lt;-</span><span class="nf">openTomahawkOutput</span><span class="p">(</span><span class="s">&quot;example.two&quot;</span><span class="p">))</span>
   <span class="n">user</span>  <span class="n">system</span> <span class="n">elapsed</span> 
  <span class="m">0.003</span>   <span class="m">0.000</span>   <span class="m">0.002</span>
</pre></div>
</td></tr></table>

<div class="admonition success">
<p class="admonition-title">Opening pre-computed LD data</p>
<p>Success: This object is now ready to be used by other downstream functions in <code>rtomahawk</code>.</p>
</div>
<p>In some situations you <em>do</em> want to load data into memory using the file handle
pointer in a <code>twk</code> object. This is trivially done with the <code>readRecords</code>
subroutine that require a <code>twk</code> class object as a required parameter. As it is
very simple to accidentally load hundreds of millions of records if you are not
careful we introduced the logical flag <code>really</code>. By default this flag is set to
<code>FALSE</code> and will force terminate the loading procedure at 10 million records.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="n">y</span><span class="o">&lt;-</span><span class="nf">readRecords</span><span class="p">(</span><span class="n">twk</span><span class="p">,</span> <span class="n">really</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span>
</pre></div>
</td></tr></table>

<p>If the safety limit is reached a warning message will be printed to the console:</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="o">&gt;</span> <span class="n">y</span><span class="o">&lt;-</span><span class="nf">readRecords</span><span class="p">(</span><span class="n">twk</span><span class="p">,</span> <span class="n">really</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span>
<span class="n">limit</span> <span class="n">reached</span>
</pre></div>
</td></tr></table>

<div class="admonition warning">
<p class="admonition-title">Source of memory inefficiency</p>
<p>In the current release, this subroutine load records (C structs) into vectors of records. This
is unlike the required column-store format used in <code>R</code>. Because of this, we perform
an implicit transposition of the internal structs into appropriate <code>std::vectors</code> followed
by convertion into the appropriate <code>data.table</code> format. During this final convertion into
<code>SEXP</code> structures, <code>Rcpp</code> will perform an unneccessary copy of the data resulting in a transient
use (spike) of excess memory. We will address this problem in upcoming releases.</p>
</div>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>&gt; y
An object of class twk
Has internal data: 10000000 records
  FLAG ridA   posA ridB   posB REFREF REFALT ALTREF ALTALT           D
1   11    6 123695    6 227960   4978     23      0      7 0.001389390
2    3    6 143500    6 228257   4425    387     78    118 0.019615739
3    3    6 144397    6 224822   4856     16     98     38 0.007295037
4    3    6 144398    6 224822   4856     16     98     38 0.007295037
5    3    6 150279    6 225420   4229    345    233    201 0.030687481
     Dprime         R        R2            P ChiSqFisher ChiSqModel
1 1.0000000 0.4819338 0.2322602 1.304179e-16   1163.1592          0
2 0.5574107 0.3359277 0.1128474 5.789946e-71    565.1399          0
3 0.6954327 0.4345684 0.1888497 1.786734e-49    945.7594          0
4 0.6954327 0.4345684 0.1888497 1.786734e-49    945.7594          0
5 0.3974391 0.3499740 0.1224818 8.508703e-90    613.3890          0
...
         FLAG ridA    posA ridB    posB REFREF REFALT ALTREF ALTALT           D
9999995     3    6 1834075    6 1888336   4932      4     52     20 0.003924711
9999996    11    6 1834098    6 1885227   4892      0     97     19 0.003706051
9999997    11    6 1834098    6 1886103   4982      0      7     19 0.003774233
9999998    11    6 1834098    6 1890443   4873      0    116     19 0.003691657
9999999    11    6 1834098    6 1890998   4873      0    116     19 0.003691657
10000000   11    6 1834098    6 1891303   4873      0    116     19 0.003691657
            Dprime         R        R2            P ChiSqFisher ChiSqModel
9999995  0.8309022 0.4774060 0.2279165 8.176940e-35   1141.4056          0
9999996  1.0000000 0.4007599 0.1606085 1.856439e-32    804.3274          0
9999997  1.0000000 0.8542505 0.7297439 4.211277e-48   3654.5574          0
9999998  1.0000000 0.3707673 0.1374684 4.173075e-31    688.4415          0
9999999  1.0000000 0.3707673 0.1374684 4.173075e-31    688.4415          0
10000000 1.0000000 0.3707673 0.1374684 4.173075e-31    688.4415          0
</pre></div>
</td></tr></table>

<p>A more useful procedure is to slice out regions of interest that match some set
of criteron. Such as for example retrieving records from the region
chr6:5e6-10e6 with an <code>R2</code> value between 0.4 and 0.8 and a P-value &lt; 0.001:</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="n">f</span><span class="o">&lt;-</span><span class="nf">setFilters</span><span class="p">(</span><span class="n">minR2</span><span class="o">=</span><span class="m">0.6</span><span class="p">,</span> <span class="n">maxR2</span><span class="o">=</span><span class="m">0.8</span><span class="p">,</span> <span class="n">maxP</span> <span class="o">=</span> <span class="m">0.001</span><span class="p">)</span>
<span class="n">recs</span><span class="o">&lt;-</span><span class="nf">readRecords</span><span class="p">(</span><span class="n">twk</span><span class="p">,</span><span class="s">&quot;6:5e6-10e6&quot;</span><span class="p">,</span><span class="n">filters</span><span class="o">=</span><span class="n">f</span><span class="p">,</span><span class="n">really</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
</pre></div>
</td></tr></table>

<div class="admonition note">
<p class="admonition-title">Interval slicing</p>
<p>Intervals are left-inclusive and right-exclusive [A, B) and use 1-base coordinates. The
interval syntax has to match either: A) <code>contig</code>, B) <code>contig:pos</code>, or C) <code>contig:pos-pos</code>.
It is important to note that interval slicing is performed on the forward position <em>only</em>
(<code>posA</code>) and that by default the output data generarted from Tomahawk have bidirectional
symmetry such that two tuples (A,B) == (B,A), exists and are mirrored.</p>
</div>
<p>We can ascertain that the filtering procedure work correctly by looking at the
summary statistics for each column and note that <code>posA</code> is bounded by 5Mb-10Mb,
<code>R2</code> is bounded by [0.6, 0.8], and <code>P</code> &lt; 0.001:</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="nf">summary</span><span class="p">(</span><span class="n">recs</span><span class="p">)</span>
</pre></div>
</td></tr></table>

<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>                     Min.      1st Qu.        Median         Mean      3rd Qu.
FLAG         3.000000e+00 3.000000e+00  3.000000e+00 5.311864e+01 3.000000e+00
ridA         3.659840e+05 3.659840e+05  3.659840e+05 3.659840e+05 3.659840e+05
<span class="hll">posA         5.000012e+06 7.093750e+06  7.887604e+06 7.780695e+06 8.673979e+06
</span>ridB         3.659840e+05 3.659840e+05  3.659840e+05 3.659840e+05 3.659840e+05
posB         2.869352e+06 7.094321e+06  7.887945e+06 7.782022e+06 8.675012e+06
REFREF       0.000000e+00 2.801000e+03  4.728000e+03 3.836738e+03 4.953000e+03
REFALT       0.000000e+00 2.000000e+00  1.800000e+01 2.416587e+02 1.300000e+02
ALTREF       0.000000e+00 3.000000e+00  2.600000e+01 2.241404e+02 1.630000e+02
ALTALT       0.000000e+00 3.500000e+01  1.630000e+02 7.054628e+02 1.076000e+03
D           -2.231411e-01 4.564901e-03  2.287482e-02 5.856900e-02 1.283542e-01
Dprime       7.749466e-01 8.886666e-01  9.506363e-01 9.368918e-01 9.961449e-01
R            7.745978e-01 8.045203e-01  8.354463e-01 8.354627e-01 8.654019e-01
<span class="hll">R2           6.000017e-01 6.472530e-01  6.979705e-01 6.992205e-01 7.489204e-01
</span><span class="hll">P            0.000000e+00 0.000000e+00 2.578244e-297 2.630748e-12 2.487910e-85
</span>ChiSqFisher  3.004809e+03 3.241443e+03  3.495436e+03 3.501696e+03 3.750593e+03
ChiSqModel   0.000000e+00 0.000000e+00  0.000000e+00 0.000000e+00 0.000000e+00
                    Max.
FLAG        2.063000e+03
ridA        3.659840e+05
<span class="hll">posA        9.999976e+06
</span>ridB        3.659840e+05
posB        1.240232e+07
REFREF      5.003000e+03
REFALT      4.945000e+03
ALTREF      5.005000e+03
ALTALT      4.999000e+03
D           2.233769e-01
Dprime      1.000000e+00
R           8.944252e-01
<span class="hll">R2          7.999965e-01
</span><span class="hll">P           2.392816e-07
</span>ChiSqFisher 4.006383e+03
ChiSqModel  0.000000e+00
</pre></div>
</td></tr></table>

<div class="admonition success">
<p class="admonition-title">Opening and slicing LD data</p>
<p>Success: This object is now ready to be used by other downstream functions in <code>rtomahawk</code>.</p>
</div>
<p>There are a large number of filter parameters available to slice out the exact
data of interest:</p>
<table>
<thead>
<tr>
<th>Field</th>
<th>Type</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>flagInclude</code></td>
<td><code>integer</code></td>
</tr>
<tr>
<td><code>flagExclude</code></td>
<td><code>integer</code></td>
</tr>
<tr>
<td><code>minR</code></td>
<td><code>numeric</code></td>
</tr>
<tr>
<td><code>maxR</code></td>
<td><code>numeric</code></td>
</tr>
<tr>
<td><code>minR2</code></td>
<td><code>numeric</code></td>
</tr>
<tr>
<td><code>maxR2</code></td>
<td><code>numeric</code></td>
</tr>
<tr>
<td><code>minD</code></td>
<td><code>numeric</code></td>
</tr>
<tr>
<td><code>maxD</code></td>
<td><code>numeric</code></td>
</tr>
<tr>
<td><code>minDprime</code></td>
<td><code>numeric</code></td>
</tr>
<tr>
<td><code>maxDprime</code></td>
<td><code>numeric</code></td>
</tr>
<tr>
<td><code>minP</code></td>
<td><code>numeric</code></td>
</tr>
<tr>
<td><code>maxP</code></td>
<td><code>numeric</code></td>
</tr>
<tr>
<td><code>minP1</code></td>
<td><code>numeric</code></td>
</tr>
<tr>
<td><code>maxP1</code></td>
<td><code>numeric</code></td>
</tr>
<tr>
<td><code>minP2</code></td>
<td><code>numeric</code></td>
</tr>
<tr>
<td><code>maxP2</code></td>
<td><code>numeric</code></td>
</tr>
<tr>
<td><code>minQ1</code></td>
<td><code>numeric</code></td>
</tr>
<tr>
<td><code>maxQ1</code></td>
<td><code>numeric</code></td>
</tr>
<tr>
<td><code>minQ2</code></td>
<td><code>numeric</code></td>
</tr>
<tr>
<td><code>maxQ2</code></td>
<td><code>numeric</code></td>
</tr>
<tr>
<td><code>minChiSqFisher</code></td>
<td><code>numeric</code></td>
</tr>
<tr>
<td><code>maxChiSqFisher</code></td>
<td><code>numeric</code></td>
</tr>
<tr>
<td><code>minChiSqModel</code></td>
<td><code>numeric</code></td>
</tr>
<tr>
<td><code>maxChiSqModel</code></td>
<td><code>numeric</code></td>
</tr>
<tr>
<td><code>upperOnly</code></td>
<td><code>logical</code></td>
</tr>
<tr>
<td><code>lowerOnly</code></td>
<td><code>logical</code></td>
</tr>
</tbody>
</table>
<p>See <code>?setFilters</code> for more information.</p>
<h2 id="calculating-linkage-disequilibrium">Calculating linkage-disequilibrium<a class="headerlink" href="#calculating-linkage-disequilibrium" title="Permanent link">&para;</a></h2>
<h2 id="plotting-data-with-rtomahawk">Plotting data with <code>rtomahawk</code><a class="headerlink" href="#plotting-data-with-rtomahawk" title="Permanent link">&para;</a></h2>
<h3 id="color-schemes">Color schemes<a class="headerlink" href="#color-schemes" title="Permanent link">&para;</a></h3>
<p><code>rtomahawk</code> comes prepacked with a number of color schemes. Most of these are
taken directly from the
<a href="https://cran.r-project.org/web/packages/viridis/"><code>viridis</code></a> <code>R</code> package and
are included here independent of that package to remove the package
interdependency. These color schemes are callable as functions taking as require
arguments the number of distinct values (<code>n</code>) and optionally the desired opacity
(alpha) level (<code>alpha</code>).</p>
<table>
<thead>
<tr>
<th>Color scheme</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>viridis</code></td>
</tr>
<tr>
<td><code>plasma</code></td>
</tr>
<tr>
<td><code>magma</code></td>
</tr>
<tr>
<td><code>inferno</code></td>
</tr>
<tr>
<td><code>cividis</code></td>
</tr>
<tr>
<td><code>default</code></td>
</tr>
</tbody>
</table>
<p>We can plot these values as a gradient of values from [0, 100) and as [0, 11] by
calling the <code>displayColors</code> function:</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="nf">displayColors</span><span class="p">()</span>
</pre></div>
</td></tr></table>

<p><img src="../images/rtwk_colors.jpeg"></p>
<h3 id="square-representation">Square representation<a class="headerlink" href="#square-representation" title="Permanent link">&para;</a></h3>
<p>Graphically representing LD data is useful for checking data quality and in
exploration. If your data has already been loaded into memory using
<code>readRecords</code> it is possible to plot this data as individual data points using
the <code>plotLD</code> function. This approach is generally only feasable for visualizing
smaller genomic regions, for example, &lt; 2-3 megabases. If you are investigating
longer ranges than this you should consider using the <a href="../aggregation/">aggregation</a> functions
provided in <code>tomahawk</code>/<code>rtomahawk</code>. </p>
<p>Because of the vast number of data points rendered and the finite amount of
pixels available, we render data points with an opacity gradient scaled
according to its <code>R2</code> value from [0.1, 1]. This allows for mixing of both colors
and opacities to more clearly represent the distribution of the underlying data.
It is possible to disable this functionality by setting the optional argument
<code>opacity</code> to <code>FALSE</code>. In the following examples, we render both a large region
(5-8 Mb) and a small region (5.0-5.6 Mb) with and without the opacity flag set.
We also showcase the different color schemes.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="c1"># Load some local data into memory.</span>
<span class="n">twk</span> <span class="o">&lt;-</span> <span class="nf">openTomahawkOutput</span><span class="p">(</span><span class="s">&quot;1kgp3_chr6.two&quot;</span><span class="p">)</span>
<span class="n">y</span> <span class="o">&lt;-</span> <span class="nf">readRecords</span><span class="p">(</span><span class="n">twk</span><span class="p">,</span><span class="s">&quot;6:5e6-10e6&quot;</span><span class="p">,</span> <span class="n">really</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
<span class="c1"># Plot two panels in the same figure.</span>
<span class="c1"># Top panel: opacity gradient from [0.1, 1.0] mapping to </span>
<span class="c1"># R2 range [0.1, 0.1, 0.2, ..., 1.0]</span>
<span class="c1"># &lt;COLOR&gt; placeholder gets replaced with one of the color </span>
<span class="c1"># schemes described below.</span>
<span class="nf">par</span><span class="p">(</span><span class="n">mfrow</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="m">1</span><span class="p">))</span>
<span class="nf">plotLD</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">5e6</span><span class="p">,</span><span class="m">8e6</span><span class="p">),</span> <span class="n">xlim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">5e6</span><span class="p">,</span><span class="m">8e6</span><span class="p">),</span> <span class="n">colors</span><span class="o">=&lt;</span><span class="n">COLOR</span><span class="o">&gt;</span><span class="p">,</span> <span class="n">bg</span><span class="o">=&lt;</span><span class="n">COLOR</span><span class="o">&gt;</span><span class="p">(</span><span class="m">11</span><span class="p">)</span><span class="n">[1]</span><span class="p">)</span>
<span class="nf">plotLD</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">5e6</span><span class="p">,</span><span class="m">8e6</span><span class="p">),</span> <span class="n">xlim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">5e6</span><span class="p">,</span><span class="m">8e6</span><span class="p">),</span> <span class="n">colors</span><span class="o">=&lt;</span><span class="n">COLOR</span><span class="o">&gt;</span><span class="p">,</span> <span class="n">bg</span><span class="o">=&lt;</span><span class="n">COLOR</span><span class="o">&gt;</span><span class="p">(</span><span class="m">11</span><span class="p">)</span><span class="n">[1]</span><span class="p">,</span> <span class="n">opacity</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span>
</pre></div>
</td></tr></table>

<table>
<thead>
<tr>
<th>Color scheme</th>
<th>Image</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>default</code></td>
<td><img src="../images/twk_plotLD_default_quad.jpeg"></td>
</tr>
<tr>
<td><code>cividis</code></td>
<td><img src="../images/twk_plotLD_cividis_quad.jpeg"></td>
</tr>
<tr>
<td><code>inferno</code></td>
<td><img src="../images/twk_plotLD_inferno_quad.jpeg"></td>
</tr>
<tr>
<td><code>magma</code></td>
<td><img src="../images/twk_plotLD_magma_quad.jpeg"></td>
</tr>
<tr>
<td><code>plasma</code></td>
<td><img src="../images/twk_plotLD_plasma_quad.jpeg"></td>
</tr>
<tr>
<td><code>viridis</code></td>
<td><img src="../images/twk_plotLD_viridis_quad.jpeg"></td>
</tr>
</tbody>
</table>
<p>By default, the birectionally symmetric output data from <code>tomahawk</code> is kept (i.e. both (A,B) and (B,A))
tuples are kept. For this reason, the output plot will be square (or rectangular with mismatched <code>xlim</code>
and <code>ylim</code> parameters). In some cases, only the upper or lower triangular values are desired. We can
control what values are plotted with the logical <code>upper</code> and <code>lower</code> parameters.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="nf">par</span><span class="p">(</span><span class="n">mfrow</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">3</span><span class="p">))</span>
<span class="nf">plotLD</span><span class="p">(</span><span class="n">y</span><span class="p">,</span><span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">5e6</span><span class="p">,</span><span class="m">8e6</span><span class="p">),</span><span class="n">xlim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">5e6</span><span class="p">,</span><span class="m">8e6</span><span class="p">),</span><span class="n">colors</span><span class="o">=</span><span class="nf">viridis</span><span class="p">(</span><span class="m">11</span><span class="p">),</span><span class="n">bg</span><span class="o">=</span><span class="nf">viridis</span><span class="p">(</span><span class="m">11</span><span class="p">)</span><span class="n">[1]</span><span class="p">)</span>
<span class="nf">plotLD</span><span class="p">(</span><span class="n">y</span><span class="p">,</span><span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">5e6</span><span class="p">,</span><span class="m">8e6</span><span class="p">),</span><span class="n">xlim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">5e6</span><span class="p">,</span><span class="m">8e6</span><span class="p">),</span><span class="n">colors</span><span class="o">=</span><span class="nf">viridis</span><span class="p">(</span><span class="m">11</span><span class="p">),</span><span class="n">bg</span><span class="o">=</span><span class="nf">viridis</span><span class="p">(</span><span class="m">11</span><span class="p">)</span><span class="n">[1]</span><span class="p">,</span> <span class="n">upper</span><span class="o">=</span><span class="bp">T</span><span class="p">)</span>
<span class="nf">plotLD</span><span class="p">(</span><span class="n">y</span><span class="p">,</span><span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">5e6</span><span class="p">,</span><span class="m">8e6</span><span class="p">),</span><span class="n">xlim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">5e6</span><span class="p">,</span><span class="m">8e6</span><span class="p">),</span><span class="n">colors</span><span class="o">=</span><span class="nf">viridis</span><span class="p">(</span><span class="m">11</span><span class="p">),</span><span class="n">bg</span><span class="o">=</span><span class="nf">viridis</span><span class="p">(</span><span class="m">11</span><span class="p">)</span><span class="n">[1]</span><span class="p">,</span> <span class="n">lower</span><span class="o">=</span><span class="bp">T</span><span class="p">)</span>
</pre></div>
</td></tr></table>

<p><img src="../images/twk_plotLD_upperlower.jpeg"></p>
<h3 id="triangular-representation">Triangular representation<a class="headerlink" href="#triangular-representation" title="Permanent link">&para;</a></h3>
<p>In many cases, there is generally no need to graphically represent the entire
square (or rectangular) symmetric matrix of associations. This is especially
true when combining multiple graphs together to create a more comprehensive
picture of a particular region or feature. The topic of combining plots will be
explored in greter detail below. All of the examples here involves the
subroutine <code>plotLDTriangular</code>.  As a design choice, we decided to restrict the
rendered y-axis data such that it is always bounded by the x-axis limits. For
this reason, these plots will always be triangular will partially "missing"
(omitted) values even if they are technically present in the dataset. </p>
<p>First, we describe its most basic use case. In the following example, we will
plot a section of associations with the <code>viridis</code> color palette and specifiy the
background to be the lowest (in this case, first) color in that scheme.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="nf">plotLDTriangular</span><span class="p">(</span><span class="n">y</span><span class="p">,</span><span class="n">colors</span><span class="o">=</span><span class="nf">viridis</span><span class="p">(</span><span class="m">11</span><span class="p">),</span><span class="n">bg</span><span class="o">=</span><span class="nf">viridis</span><span class="p">(</span><span class="m">11</span><span class="p">)</span><span class="n">[1]</span><span class="p">)</span>
</pre></div>
</td></tr></table>

<p><img src="../images/twk_plotLD_triangular.jpeg"></p>
<p>Because of the smallish haplotype blocks in humans, most of these visible
triangular structures will have limited span in the y-axis. We can truncate the
y-axis to zoom into the local neighbourhood and more accurately display the
local haplotype structure.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="nf">plotLDTriangular</span><span class="p">(</span><span class="n">y</span><span class="p">,</span><span class="n">colors</span><span class="o">=</span><span class="nf">viridis</span><span class="p">(</span><span class="m">11</span><span class="p">),</span><span class="n">bg</span><span class="o">=</span><span class="nf">viridis</span><span class="p">(</span><span class="m">11</span><span class="p">)</span><span class="n">[1]</span><span class="p">,</span><span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">300e3</span><span class="p">))</span>
</pre></div>
</td></tr></table>

<p><img src="../images/twk_plotLD_triangular_truncate.jpeg"></p>
<p>It is possible to control the orientation (rotation) of the output graph by
specifying the <code>orientation</code> parameter. The numerial encodings are: 1)
standard; 2) upside down;
3) left-right flipped; and 4) right-left flipped.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="nf">par</span><span class="p">(</span><span class="n">mfrow</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="m">2</span><span class="p">))</span>
<span class="nf">plotLDTriangular</span><span class="p">(</span><span class="n">y</span><span class="p">,</span><span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">300e3</span><span class="p">),</span><span class="n">colors</span><span class="o">=</span><span class="nf">viridis</span><span class="p">(</span><span class="m">11</span><span class="p">),</span><span class="n">bg</span><span class="o">=</span><span class="nf">viridis</span><span class="p">(</span><span class="m">11</span><span class="p">)</span><span class="n">[1]</span><span class="p">,</span> <span class="n">orientation</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span>
<span class="nf">plotLDTriangular</span><span class="p">(</span><span class="n">y</span><span class="p">,</span><span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">300e3</span><span class="p">),</span><span class="n">colors</span><span class="o">=</span><span class="nf">viridis</span><span class="p">(</span><span class="m">11</span><span class="p">),</span><span class="n">bg</span><span class="o">=</span><span class="nf">viridis</span><span class="p">(</span><span class="m">11</span><span class="p">)</span><span class="n">[1]</span><span class="p">,</span> <span class="n">orientation</span> <span class="o">=</span> <span class="m">2</span><span class="p">)</span>
<span class="nf">plotLDTriangular</span><span class="p">(</span><span class="n">y</span><span class="p">,</span><span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">300e3</span><span class="p">),</span><span class="n">colors</span><span class="o">=</span><span class="nf">viridis</span><span class="p">(</span><span class="m">11</span><span class="p">),</span><span class="n">bg</span><span class="o">=</span><span class="nf">viridis</span><span class="p">(</span><span class="m">11</span><span class="p">)</span><span class="n">[1]</span><span class="p">,</span> <span class="n">orientation</span> <span class="o">=</span> <span class="m">3</span><span class="p">)</span>
<span class="nf">plotLDTriangular</span><span class="p">(</span><span class="n">y</span><span class="p">,</span><span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">300e3</span><span class="p">),</span><span class="n">colors</span><span class="o">=</span><span class="nf">viridis</span><span class="p">(</span><span class="m">11</span><span class="p">),</span><span class="n">bg</span><span class="o">=</span><span class="nf">viridis</span><span class="p">(</span><span class="m">11</span><span class="p">)</span><span class="n">[1]</span><span class="p">,</span> <span class="n">orientation</span> <span class="o">=</span> <span class="m">4</span><span class="p">)</span>
</pre></div>
</td></tr></table>

<p><img src="../images/twk_plotLD_triangular_orientations.jpeg"></p>
<p>It is possible to completely disable all anotation by setting the logical
parameter <code>annotate</code> to <code>FALSE</code>. This will remove titles, axes, and ticks.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="nf">plotLDTriangular</span><span class="p">(</span><span class="n">y</span><span class="p">,</span><span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">300e3</span><span class="p">),</span><span class="n">colors</span><span class="o">=</span><span class="nf">viridis</span><span class="p">(</span><span class="m">11</span><span class="p">),</span><span class="n">bg</span><span class="o">=</span><span class="nf">viridis</span><span class="p">(</span><span class="m">11</span><span class="p">)</span><span class="n">[1]</span><span class="p">,</span> <span class="n">orientation</span> <span class="o">=</span> <span class="m">1</span><span class="p">,</span> <span class="n">annotate</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>
</pre></div>
</td></tr></table>

<p><img src="../images/twk_plotLD_triangular_orientations_no_annotation.jpeg"></p>
<p>Note that these plotting functions respect the global <code>mar</code> (margin) values and
(by default) will have white space around it. We can change the global <code>par</code>
argument to, for example, zero to remove these margins when annotation is
disabled for edge-to-edge graphics.</p>
<p><img src="../images/twk_plotLD_triangular_orientations_no_annotation_nomar.jpeg"></p>
<h3 id="visualizing-gwas-data-and-ld">Visualizing GWAS data and LD<a class="headerlink" href="#visualizing-gwas-data-and-ld" title="Permanent link">&para;</a></h3>
<p>In this section we will produce LocusZoom-like plots using
<a href="https://en.wikipedia.org/wiki/Genome-wide_association_study">GWAS</a> data from
the <a href="www.ukbiobank.ac.uk/">UK BioBank</a> comprised exclusively of white British
individuals. <a href="www.ed.ac.uk/roslin">The Roslin Institute</a> at the <a href="www.ed.ac.uk/">University of
Edinbrugh</a> host a data browser of associatins called <a href="http://geneatlas.roslin.ed.ac.uk/">Gene
Atlas</a>. In the following examples we will
investigate the association of genotypes at chromosome 6 and diabetes in this
cohort. Data in its entirety can be <a href="http://geneatlas.roslin.ed.ac.uk/downloads/?traits=493">explored
further</a> using the Gene
Atlas. To reproduce the results below download the
<a href="http://static.geneatlas.roslin.ed.ac.uk/gwas/allWhites/imputed/data.copy/imputed.allWhites.selfReported_n_1245.chr6.csv.gz">imputed</a>
data for chromosome 6 and the associated <a href="http://static.geneatlas.roslin.ed.ac.uk/gwas/allWhites/snps/extended/snps.imputed.chr6.csv.gz">positional
information</a>.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="c1"># Downloaded data from:</span>
<span class="c1"># http://static.geneatlas.roslin.ed.ac.uk/gwas/allWhites/imputed/data.copy/imputed.allWhites.selfReported_n_1245.chr6.csv.gz</span>
<span class="c1"># http://static.geneatlas.roslin.ed.ac.uk/gwas/allWhites/snps/extended/snps.imputed.chr6.csv.gz</span>
<span class="nf">library</span><span class="p">(</span><span class="n">data.table</span><span class="p">)</span> <span class="c1"># For speedier reading of data.</span>
<span class="c1"># We zcat (uncompress gzipped archive) directly into fread.</span>
<span class="c1"># This may not work for Windows users. In that case, decompress</span>
<span class="c1"># the files manually first.</span>
<span class="n">x</span> <span class="o">&lt;-</span> <span class="nf">fread</span><span class="p">(</span><span class="s">&quot;zcat imputed.allWhites.selfReported_n_1245.chr6.csv.gz&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot; &quot;</span><span class="p">)</span>
<span class="n">snp</span> <span class="o">&lt;-</span> <span class="nf">fread</span><span class="p">(</span><span class="s">&quot;zcat snps.imputed.chr6.csv.gz&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot; &quot;</span><span class="p">)</span>
<span class="c1"># Keep matching data available in both files. We neeed to maintain </span>
<span class="c1"># parity between the two sets.</span>
<span class="n">snp</span> <span class="o">&lt;-</span> <span class="n">snp</span><span class="nf">[match</span><span class="p">(</span><span class="n">x</span><span class="o">$</span><span class="n">SNP</span><span class="p">,</span> <span class="n">snp</span><span class="o">$</span><span class="n">SNP</span><span class="p">),</span><span class="n">]</span>
<span class="c1"># Transform linear-scale (untransformed) P-values into -log10(P).</span>
<span class="n">snp</span><span class="o">$</span><span class="n">p</span> <span class="o">&lt;-</span> <span class="o">-</span><span class="nf">log10</span><span class="p">(</span><span class="n">x</span><span class="o">$</span><span class="n">`PV-selfReported_n_1245`</span><span class="p">)</span>

<span class="c1"># Setup path to local Tomahawk file to compute LD against.</span>
<span class="n">twk2</span> <span class="o">&lt;-</span> <span class="nf">new</span><span class="p">(</span><span class="s">&quot;twk&quot;</span><span class="p">)</span>
<span class="n">twk2</span><span class="o">@</span><span class="n">file.path</span> <span class="o">&lt;-</span> <span class="s">&quot;1kgp3_chr6.twk&quot;</span>

<span class="c1"># Load human recombination data (hg19) supplied with `rtomahawk`</span>
<span class="nf">data</span><span class="p">(</span><span class="n">gmap</span><span class="p">)</span>
<span class="c1"># Region of 1 Mb in either direction of target.</span>
<span class="n">single</span> <span class="o">&lt;-</span> <span class="nf">plotLZ</span><span class="p">(</span><span class="n">twk2</span><span class="p">,</span> <span class="s">&quot;6:20694884&quot;</span><span class="p">,</span> <span class="n">snp</span><span class="p">,</span> <span class="n">gmap</span><span class="p">,</span> <span class="n">window</span><span class="o">=</span><span class="m">1e6</span><span class="p">,</span> <span class="n">minR2</span><span class="o">=</span><span class="m">0</span><span class="p">)</span>
<span class="c1"># Region of 50 Kb in either direction of target.</span>
<span class="n">single</span> <span class="o">&lt;-</span> <span class="nf">plotLZ</span><span class="p">(</span><span class="n">twk2</span><span class="p">,</span> <span class="s">&quot;6:20694884&quot;</span><span class="p">,</span> <span class="n">snp</span><span class="p">,</span> <span class="n">gmap</span><span class="p">,</span> <span class="n">window</span><span class="o">=</span><span class="m">50e3</span><span class="p">,</span> <span class="n">minR2</span><span class="o">=</span><span class="m">0</span><span class="p">)</span>
</pre></div>
</td></tr></table>

<p><img src="../images/twk_locuszoom.jpeg">
<img src="../images/twk_locuszoom_50k.jpeg"></p>
<div class="admonition note">
<p class="admonition-title">Internal LD computation</p>
<p>The <code>plotLZ</code> plotting function will internally compute linkage-disequilibrium
for a target SNV and it's surrounding genomic region. This background computation 
is stored in a temporary file and then loaded back into memory and returned as a new
<code>twk</code> class instance. If you have further use of this information then you
need to capture the return value of this function. It is possible to change the
default <code>temp</code> directory used in <code>R</code> with the subroutine <code>set.tempdir</code>. Please note
that all data written to this temporary directory will be purged upon exiting the
current session of <code>R</code>.</p>
</div>
<p>These functions are extremely fast as the R-bindings use <code>.Call</code> commands to
communicate with the compiled C++ shared object:</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5
6
7</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="o">&gt;</span> <span class="nf">system.time</span><span class="p">(</span><span class="nf">plotLZ</span><span class="p">(</span><span class="n">twk2</span><span class="p">,</span> <span class="s">&quot;6:20694884&quot;</span><span class="p">,</span> <span class="n">snp</span><span class="p">,</span> <span class="n">gmap</span><span class="p">,</span> <span class="n">window</span><span class="o">=</span><span class="m">1e6</span><span class="p">,</span> <span class="n">minR2</span><span class="o">=</span><span class="m">0</span><span class="p">))</span>
   <span class="n">user</span>  <span class="n">system</span> <span class="n">elapsed</span> 
  <span class="m">1.405</span>   <span class="m">0.001</span>   <span class="m">1.407</span>

<span class="o">&gt;</span> <span class="nf">system.time</span><span class="p">(</span><span class="nf">plotLZ</span><span class="p">(</span><span class="n">twk2</span><span class="p">,</span> <span class="s">&quot;6:20694884&quot;</span><span class="p">,</span> <span class="n">snp</span><span class="p">,</span> <span class="n">gmap</span><span class="p">,</span> <span class="n">window</span><span class="o">=</span><span class="m">50e3</span><span class="p">,</span> <span class="n">minR2</span><span class="o">=</span><span class="m">0</span><span class="p">))</span>
   <span class="n">user</span>  <span class="n">system</span> <span class="n">elapsed</span> 
  <span class="m">0.115</span>   <span class="m">0.004</span>   <span class="m">0.119</span>
</pre></div>
</td></tr></table>

<p>Almost all of this time is spent rendering symbols in the vectorized plot. The
same command using the CLI takes roughly 1/3rd of the time:</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>$ <span class="nb">time</span> tomahawk scalc -i 1kgp3_chr6.twk -I <span class="m">6</span>:20694884 -w <span class="m">1000000</span>  &gt; /dev/null
real  0m0.452s
user  0m0.937s
sys   0m0.061s
</pre></div>
</td></tr></table>

<p>Many aspects of the plots can be customized to your needs. For example,
modifying the datapoint symbol <code>pch</code> in the valid range [21, 25]. Note that all
other <code>pch</code> values are technically valid but will not generate a fill color.</p>
<p><img src="../images/twk_locuszoom_pch.jpeg"></p>
<p>If your visualizations require additional data layers it is possible to combine
these into a single plot as <code>rtomahawk</code> renders plots using base-R. In this
example we will combine two plots generated by <code>rtomahawk</code>: the GWAS P-value and
its single-site LD together with the all-vs-all pairwise LD for the same region.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5
6</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="nf">par</span><span class="p">(</span><span class="n">mfrow</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="m">1</span><span class="p">),</span> <span class="n">mar</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">5</span><span class="p">,</span><span class="m">3</span><span class="p">,</span><span class="m">5</span><span class="p">))</span> <span class="c1"># Set bottom margin to 0.</span>
<span class="n">single</span> <span class="o">&lt;-</span> <span class="nf">plotLZ</span><span class="p">(</span><span class="n">twk2</span><span class="p">,</span> <span class="s">&quot;6:20694884&quot;</span><span class="p">,</span> <span class="n">snp</span><span class="p">,</span> <span class="n">gmap</span><span class="p">,</span> <span class="n">window</span><span class="o">=</span><span class="m">1e6</span><span class="p">,</span> <span class="n">minR2</span><span class="o">=</span><span class="m">0</span><span class="p">,</span> <span class="n">xlab</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">,</span> <span class="n">xaxt</span><span class="o">=</span><span class="s">&quot;n&quot;</span><span class="p">)</span>
<span class="n">twk</span> <span class="o">&lt;-</span> <span class="nf">openTomahawkOutput</span><span class="p">(</span><span class="s">&quot;test_region.two&quot;</span><span class="p">)</span>
<span class="n">y</span> <span class="o">&lt;-</span> <span class="nf">readRecords</span><span class="p">(</span><span class="n">twk</span><span class="p">,</span> <span class="n">really</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
<span class="nf">par</span><span class="p">(</span><span class="n">mar</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">5</span><span class="p">,</span><span class="m">5</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="m">5</span><span class="p">))</span> <span class="c1"># Set top margin to 0.</span>
<span class="nf">plotLDTriangular</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">500e3</span><span class="p">),</span> <span class="n">xlim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">20694884</span> <span class="o">-</span> <span class="m">1e6</span><span class="p">,</span> <span class="m">20694884</span> <span class="o">+</span> <span class="m">1e6</span><span class="p">),</span> <span class="n">colors</span><span class="o">=</span><span class="nf">viridis</span><span class="p">(</span><span class="m">11</span><span class="p">),</span> <span class="n">bg</span><span class="o">=</span><span class="nf">viridis</span><span class="p">(</span><span class="m">11</span><span class="p">)</span><span class="n">[1]</span><span class="p">,</span> <span class="n">orientation</span> <span class="o">=</span> <span class="m">2</span><span class="p">,</span> <span class="n">cex</span><span class="o">=</span><span class="m">.25</span><span class="p">,</span> <span class="n">main</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">)</span>
</pre></div>
</td></tr></table>

<p><img src="../images/twk_locuszoom_combine.jpeg"></p>
<p>We can zoom into the local region (100kb flanking region) by simply changing the
window parameter in <code>plotLZ</code> and the <code>xlim</code> range in <code>plotLDTriangular</code>. </p>
<p><img src="../images/twk_locuszoom_combine_100k.jpeg"></p>
<p>It is possible to add more advanced data layers using external packages with
some simple manipulations. In this example we will add a gene track using
genetic information extracted from <code>biomaRt</code> and drawn using <code>Sushi</code>, both
third-party packages.</p>
<div class="admonition note">
<p class="admonition-title">biomaRt information</p>
<p>The <code>R</code> package <code>biomaRt</code> <em>requires</em> internet connectivity to function as the
package itself is a wrapper around the Ensembl <a href="https://www.ensembl.org/biomart/martview/">BioMart</a>
tool that allows extraction of connected data from databases without having
to perform explicit programming.</p>
</div>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">biomaRt</span><span class="p">)</span>
<span class="n">mart</span> <span class="o">&lt;-</span> <span class="nf">useMart</span><span class="p">(</span><span class="n">host</span><span class="o">=</span><span class="s">&#39;http://grch37.ensembl.org&#39;</span><span class="p">,</span><span class="n">biomart</span><span class="o">=</span><span class="s">&quot;ensembl&quot;</span><span class="p">,</span> <span class="n">dataset</span><span class="o">=</span><span class="s">&quot;hsapiens_gene_ensembl&quot;</span><span class="p">)</span>
<span class="c1"># We will retrieve results from biomaRt twice because the database</span>
<span class="c1"># does _not_ allow simultaneous queries for exon-level data and</span>
<span class="c1"># gene-level data. To overcome this problem, we will perform two</span>
<span class="c1"># quries to the database and merge the results together.</span>
<span class="n">results</span> <span class="o">&lt;-</span> <span class="nf">getBM</span><span class="p">(</span><span class="n">attributes</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;ensembl_exon_id&quot;</span><span class="p">,</span>
                    <span class="s">&quot;exon_chrom_start&quot;</span><span class="p">,</span><span class="s">&quot;exon_chrom_end&quot;</span><span class="p">),</span>
                 <span class="n">filters</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;chromosome_name&quot;</span><span class="p">,</span> <span class="s">&quot;start&quot;</span><span class="p">,</span> <span class="s">&quot;end&quot;</span><span class="p">),</span> 
                 <span class="n">values</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="m">6</span><span class="p">,</span> <span class="m">20694884-1e6</span><span class="p">,</span> <span class="m">20694884+1e6</span><span class="p">),</span>
                 <span class="n">mart</span><span class="o">=</span><span class="n">mart</span><span class="p">)</span>

<span class="n">results2</span> <span class="o">&lt;-</span> <span class="nf">getBM</span><span class="p">(</span><span class="n">attributes</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;ensembl_exon_id&quot;</span><span class="p">,</span><span class="s">&quot;hgnc_symbol&quot;</span><span class="p">,</span> <span class="s">&quot;chromosome_name&quot;</span><span class="p">,</span> 
                                 <span class="s">&quot;start_position&quot;</span><span class="p">,</span> <span class="s">&quot;end_position&quot;</span><span class="p">,</span><span class="s">&quot;strand&quot;</span><span class="p">),</span>
                 <span class="n">filters</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;chromosome_name&quot;</span><span class="p">,</span> <span class="s">&quot;start&quot;</span><span class="p">,</span> <span class="s">&quot;end&quot;</span><span class="p">),</span> 
                 <span class="n">values</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="m">6</span><span class="p">,</span> <span class="m">20694884-1e6</span><span class="p">,</span> <span class="m">20694884+1e6</span><span class="p">),</span>
                 <span class="n">mart</span><span class="o">=</span><span class="n">mart</span><span class="p">)</span>

<span class="n">results</span><span class="o">&lt;-</span><span class="nf">cbind</span><span class="p">(</span><span class="n">results</span><span class="p">,</span><span class="n">results2[</span><span class="p">,</span><span class="m">-1</span><span class="p">,</span><span class="n">drop</span><span class="o">=</span><span class="bp">F</span><span class="n">]</span><span class="p">)</span>
<span class="n">results</span><span class="o">&lt;-</span><span class="n">results[results</span><span class="o">$</span><span class="n">hgnc_symbol</span><span class="o">!=</span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="n">]</span>
<span class="n">results</span><span class="o">&lt;-</span><span class="n">results[</span><span class="p">,</span><span class="nf">c</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="m">3</span><span class="p">,</span><span class="m">4</span><span class="p">,</span><span class="m">5</span><span class="p">,</span><span class="m">8</span><span class="p">)</span><span class="n">]</span>
<span class="nf">names</span><span class="p">(</span><span class="n">results</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;start&quot;</span><span class="p">,</span><span class="s">&quot;stop&quot;</span><span class="p">,</span><span class="s">&quot;gene&quot;</span><span class="p">,</span><span class="s">&quot;chrom&quot;</span><span class="p">,</span><span class="s">&quot;strand&quot;</span><span class="p">)</span>
<span class="n">results</span><span class="o">$</span><span class="n">score</span> <span class="o">=</span> <span class="s">&quot;.&quot;</span>
<span class="n">results</span><span class="o">&lt;-</span><span class="n">results[</span><span class="p">,</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;chrom&quot;</span><span class="p">,</span><span class="s">&quot;start&quot;</span><span class="p">,</span><span class="s">&quot;stop&quot;</span><span class="p">,</span><span class="s">&quot;gene&quot;</span><span class="p">,</span><span class="s">&quot;score&quot;</span><span class="p">,</span> <span class="s">&quot;strand&quot;</span><span class="p">)</span><span class="n">]</span>

<span class="nf">library</span><span class="p">(</span><span class="n">Sushi</span><span class="p">)</span>
<span class="n">chrom</span> <span class="o">=</span> <span class="m">6</span>
<span class="n">chromstart</span> <span class="o">=</span> <span class="m">20694884-1e6</span>
<span class="n">chromend</span> <span class="o">=</span> <span class="m">20694884+1e6</span>

<span class="nf">layout</span><span class="p">(</span><span class="nf">matrix</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">3</span><span class="p">,</span><span class="n">ncol</span><span class="o">=</span><span class="m">1</span><span class="p">),</span><span class="n">heights</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="m">2</span><span class="p">,</span><span class="m">1</span><span class="p">))</span>
<span class="nf">par</span><span class="p">(</span><span class="n">mar</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">5</span><span class="p">,</span><span class="m">3</span><span class="p">,</span><span class="m">5</span><span class="p">))</span> <span class="c1"># Set bottom margin to 0.</span>
<span class="n">single</span> <span class="o">&lt;-</span> <span class="nf">plotLZ</span><span class="p">(</span><span class="n">twk2</span><span class="p">,</span> <span class="s">&quot;6:20694884&quot;</span><span class="p">,</span> <span class="n">snp</span><span class="p">,</span> <span class="n">gmap</span><span class="p">,</span> <span class="n">window</span><span class="o">=</span><span class="m">1e6</span><span class="p">,</span> <span class="n">minR2</span><span class="o">=</span><span class="m">0</span><span class="p">,</span> <span class="n">xlab</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">,</span> <span class="n">xaxt</span><span class="o">=</span><span class="s">&quot;n&quot;</span><span class="p">)</span>
<span class="n">twk</span><span class="o">&lt;-</span><span class="nf">openTomahawkOutput</span><span class="p">(</span><span class="s">&quot;test_region.two&quot;</span><span class="p">)</span>
<span class="n">y</span><span class="o">&lt;-</span><span class="nf">readRecords</span><span class="p">(</span><span class="n">twk</span><span class="p">,</span><span class="n">really</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
<span class="nf">par</span><span class="p">(</span><span class="n">mar</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">5</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="m">5</span><span class="p">))</span> <span class="c1"># Set top and bottom margins to 0.</span>
<span class="nf">plotLDTriangular</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">500e3</span><span class="p">),</span> <span class="n">xlim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">20694884</span> <span class="o">-</span> <span class="m">1e6</span><span class="p">,</span> <span class="m">20694884</span> <span class="o">+</span> <span class="m">1e6</span><span class="p">),</span> <span class="n">colors</span><span class="o">=</span><span class="nf">viridis</span><span class="p">(</span><span class="m">11</span><span class="p">),</span> <span class="n">bg</span><span class="o">=</span><span class="nf">viridis</span><span class="p">(</span><span class="m">11</span><span class="p">)</span><span class="n">[1]</span><span class="p">,</span> <span class="n">orientation</span> <span class="o">=</span> <span class="m">2</span><span class="p">,</span> <span class="n">cex</span><span class="o">=</span><span class="m">.25</span><span class="p">,</span> <span class="n">main</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">)</span>
<span class="nf">par</span><span class="p">(</span><span class="n">mar</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="m">5</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="m">5</span><span class="p">))</span>
<span class="nf">plotGenes</span><span class="p">(</span><span class="n">geneinfo</span><span class="o">=</span><span class="n">results</span><span class="p">,</span> <span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">,</span> <span class="n">chromstart</span><span class="o">=</span><span class="n">chromstart</span><span class="p">,</span> <span class="n">chromend</span><span class="o">=</span><span class="n">chromend</span><span class="p">,</span> <span class="n">labeloffset</span><span class="o">=</span><span class="m">.5</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="m">1</span><span class="p">,</span> <span class="n">arrowlength</span> <span class="o">=</span> <span class="m">0.025</span><span class="p">)</span>
<span class="nf">abline</span><span class="p">(</span><span class="n">v</span><span class="o">=</span><span class="m">20694884</span><span class="p">,</span><span class="n">lty</span><span class="o">=</span><span class="s">&quot;dashed&quot;</span><span class="p">,</span><span class="n">col</span><span class="o">=</span><span class="s">&quot;grey&quot;</span><span class="p">)</span>
<span class="c1"># Highjack internal `rtomahawk` function for drawing genomic axes.</span>
<span class="c1"># This function is unexported and is not intended for general use.</span>
<span class="n">rtomahawk</span><span class="o">:::</span><span class="nf">addGenomicAxis</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="n">chromstart</span><span class="p">,</span><span class="n">chromend</span><span class="p">),</span><span class="n">at</span> <span class="o">=</span> <span class="m">1</span><span class="p">,</span> <span class="n">las</span> <span class="o">=</span> <span class="m">1</span><span class="p">,</span> <span class="bp">F</span><span class="p">)</span>
</pre></div>
</td></tr></table>

<p><img src="../images/twk_locuszoom_combine_genes.jpeg"></p>
<p>Again, we can zoom into the local region (200kb flanking region) by simply
changing the appropriate parameters in the code above.</p>
<p><img src="../images/twk_locuszoom_combine_genes_zoom.jpeg"  ></p>
<p>The plotting time can vary from milliseconds to several seconds depending on the
number of points that are being computed directly (top panel) and the number of
points that are loaded and rendered (middle panel). In the first example above,
<code>rtomahawk</code> and <code>tomahawk</code> computes and plots millions of LD associations and
data points in a few seconds on a single thread.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="o">&gt;</span> <span class="nf">system.time</span><span class="p">(</span><span class="nf">f</span><span class="p">())</span>
   <span class="n">user</span>  <span class="n">system</span> <span class="n">elapsed</span> 
  <span class="m">6.336</span>   <span class="m">0.260</span>   <span class="m">6.233</span> 
</pre></div>
</td></tr></table>

<p>In this code snippet, the function <code>f()</code>, is simply a placeholder for the code
above.</p>
<h2 id="aggregating-and-visualizing-datasets">Aggregating and visualizing datasets<a class="headerlink" href="#aggregating-and-visualizing-datasets" title="Permanent link">&para;</a></h2>
<p>Quantile-normalized (left) or linear range (right)</p>
<div class="admonition warning">
<p class="admonition-title">Unsafe method</p>
<p>In the current release of <code>rtomahawk</code>, this subroutine does <em>not</em> check for
user-interruption (for example <code>Ctrl+C</code> or <code>Ctrl+Z</code>) commands. This means that
you need to wait until the underlying process has finished or you have to
terminate the host <code>R</code> session, in turn killing the spawned process. This will
be fixed in upcoming releases.</p>
</div>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="n">twk</span><span class="o">&lt;-</span><span class="nf">openTomahawkOutput</span><span class="p">(</span><span class="s">&quot;example.two&quot;</span><span class="p">)</span>
<span class="n">x</span><span class="o">&lt;-</span><span class="nf">aggregate</span><span class="p">(</span><span class="n">twk</span><span class="p">,</span><span class="n">aggregation</span><span class="o">=</span><span class="s">&quot;r2&quot;</span><span class="p">,</span> <span class="n">reduction</span><span class="o">=</span><span class="s">&quot;count&quot;</span><span class="p">,</span><span class="n">xbins</span><span class="o">=</span><span class="m">1000</span><span class="p">,</span> <span class="n">ybins</span><span class="o">=</span><span class="m">1000</span><span class="p">,</span><span class="n">minCount</span><span class="o">=</span><span class="m">50</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">T</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="m">8</span><span class="p">)</span>
<span class="nf">par</span><span class="p">(</span><span class="n">mar</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">5</span><span class="p">,</span><span class="m">5</span><span class="p">,</span><span class="m">5</span><span class="p">,</span><span class="m">8</span><span class="p">),</span> <span class="n">mfrow</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">2</span><span class="p">))</span>
<span class="nf">plotAggregation</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">normalize</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
<span class="nf">plotAggregation</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">normalize</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>
</pre></div>
</td></tr></table>

<p><img src="../images/twk_aggregate_r2.jpeg"></p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="o">&gt;</span> <span class="n">agg</span> <span class="o">&lt;-</span> <span class="nf">aggregate</span><span class="p">(</span><span class="n">twk</span><span class="p">,</span><span class="s">&quot;r2&quot;</span><span class="p">,</span> <span class="s">&quot;count&quot;</span><span class="p">,</span> <span class="m">1000</span><span class="p">,</span> <span class="m">1000</span><span class="p">,</span> <span class="m">50</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">T</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="m">6</span><span class="p">)</span>
<span class="n">Program</span><span class="o">:</span>   <span class="n">tomahawk</span><span class="m">-062954</span><span class="nf">eb </span><span class="p">(</span><span class="n">Tools</span> <span class="n">for</span> <span class="n">computing</span><span class="p">,</span> <span class="n">querying</span> <span class="n">and</span> <span class="n">storing</span> <span class="n">LD</span> <span class="n">data</span><span class="p">)</span>
<span class="n">Libraries</span><span class="o">:</span> <span class="n">tomahawk</span><span class="m">-0.7.0</span><span class="p">;</span> <span class="n">ZSTD</span><span class="m">-1.3.1</span><span class="p">;</span> <span class="n">htslib</span> <span class="m">1.9</span>
<span class="n">Contact</span><span class="o">:</span> <span class="n">Marcus</span> <span class="n">D.</span> <span class="n">R.</span> <span class="n">Klarqvist</span> <span class="o">&lt;</span><span class="n">mk819</span><span class="o">@</span><span class="n">cam.ac.uk</span><span class="o">&gt;</span>
<span class="n">Documentation</span><span class="o">:</span> <span class="n">https</span><span class="o">://</span><span class="n">github.com</span><span class="o">/</span><span class="n">mklarqvist</span><span class="o">/</span><span class="n">tomahawk</span>
<span class="n">License</span><span class="o">:</span> <span class="n">MIT</span>
<span class="o">----------</span>
<span class="n">[2019</span><span class="m">-01-19</span> <span class="m">11</span><span class="o">:</span><span class="m">05</span><span class="o">:</span><span class="m">37</span><span class="p">,</span><span class="m">072</span><span class="n">][LOG]</span> <span class="n">Calling</span> <span class="n">aggregate...</span>
<span class="n">[2019</span><span class="m">-01-19</span> <span class="m">11</span><span class="o">:</span><span class="m">05</span><span class="o">:</span><span class="m">37</span><span class="p">,</span><span class="m">072</span><span class="n">][LOG]</span> <span class="n">Performing</span> <span class="m">2</span><span class="o">-</span><span class="n">pass</span> <span class="n">over</span> <span class="n">data...</span>
<span class="n">[2019</span><span class="m">-01-19</span> <span class="m">11</span><span class="o">:</span><span class="m">05</span><span class="o">:</span><span class="m">37</span><span class="p">,</span><span class="m">072</span><span class="n">][LOG]</span> <span class="o">=====</span> <span class="n">First</span> <span class="nf">pass </span><span class="p">(</span><span class="n">peeking</span> <span class="n">at</span> <span class="n">landscape</span><span class="p">)</span> <span class="o">=====</span>
<span class="n">[2019</span><span class="m">-01-19</span> <span class="m">11</span><span class="o">:</span><span class="m">05</span><span class="o">:</span><span class="m">37</span><span class="p">,</span><span class="m">072</span><span class="n">][LOG]</span> <span class="n">Blocks</span><span class="o">:</span> <span class="m">32</span>
<span class="n">[2019</span><span class="m">-01-19</span> <span class="m">11</span><span class="o">:</span><span class="m">05</span><span class="o">:</span><span class="m">37</span><span class="p">,</span><span class="m">072</span><span class="n">][LOG]</span> <span class="n">Uncompressed</span> <span class="n">size</span><span class="o">:</span> <span class="m">5.286261</span> <span class="n">Gb</span>
<span class="n">[2019</span><span class="m">-01-19</span> <span class="m">11</span><span class="o">:</span><span class="m">05</span><span class="o">:</span><span class="m">37</span><span class="p">,</span><span class="m">072</span><span class="n">][LOG][THREAD]</span> <span class="n">Data</span><span class="o">/</span><span class="n">thread</span><span class="o">:</span> <span class="m">881.043564</span> <span class="n">Mb</span>
<span class="n">[2019</span><span class="m">-01-19</span> <span class="m">11</span><span class="o">:</span><span class="m">05</span><span class="o">:</span><span class="m">41</span><span class="p">,</span><span class="m">040</span><span class="n">][LOG]</span> <span class="o">=====</span> <span class="n">Second</span> <span class="nf">pass </span><span class="p">(</span><span class="n">building</span> <span class="n">matrix</span><span class="p">)</span> <span class="o">=====</span>
<span class="n">[2019</span><span class="m">-01-19</span> <span class="m">11</span><span class="o">:</span><span class="m">05</span><span class="o">:</span><span class="m">41</span><span class="p">,</span><span class="m">040</span><span class="n">][LOG]</span> <span class="n">Aggregating</span> <span class="m">49</span><span class="p">,</span><span class="m">870</span><span class="p">,</span><span class="m">388</span> <span class="n">records...</span>
<span class="n">[2019</span><span class="m">-01-19</span> <span class="m">11</span><span class="o">:</span><span class="m">05</span><span class="o">:</span><span class="m">41</span><span class="p">,</span><span class="m">040</span><span class="n">][LOG][THREAD]</span> <span class="n">Allocating</span><span class="o">:</span> <span class="m">240.000000</span> <span class="n">Mb</span> <span class="n">for</span> <span class="n">matrices...</span>
<span class="n">[2019</span><span class="m">-01-19</span> <span class="m">11</span><span class="o">:</span><span class="m">05</span><span class="o">:</span><span class="m">44</span><span class="p">,</span><span class="m">797</span><span class="n">][LOG]</span> <span class="n">Aggregated</span> <span class="m">49</span><span class="p">,</span><span class="m">870</span><span class="p">,</span><span class="m">388</span> <span class="n">records</span> <span class="n">in</span> <span class="m">1</span><span class="p">,</span><span class="m">000</span><span class="p">,</span><span class="m">000</span> <span class="n">bins.</span>
<span class="n">[2019</span><span class="m">-01-19</span> <span class="m">11</span><span class="o">:</span><span class="m">05</span><span class="o">:</span><span class="m">44</span><span class="p">,</span><span class="m">798</span><span class="n">][LOG]</span> <span class="n">Finished.</span>
</pre></div>
</td></tr></table>

<p>Load a pre-computed <code>tomahawk</code> aggregation object:</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="n">agg</span> <span class="o">&lt;-</span> <span class="nf">loadAggregate</span><span class="p">(</span><span class="s">&quot;agg.twa&quot;</span><span class="p">)</span>
</pre></div>
</td></tr></table>
                
                  
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href="../tutorial/" title="CLI" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                CLI
              </span>
            </div>
          </a>
        
        
          <a href="../docker-tutorial/" title="Docker" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Docker
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            Copyright (C) 2016-present Genome Research Ltd., University of Cambridge, and Marcus D. R. Klarqvist
          </div>
        
        powered by
        <a href="https://www.mkdocs.org">MkDocs</a>
        and
        <a href="https://squidfunk.github.io/mkdocs-material/">
          Material for MkDocs</a>
      </div>
      
        
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../assets/javascripts/application.b3a2063f.js"></script>
      
      <script>app.initialize({version:"1.0.4",url:{base:".."}})</script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML"></script>
      
    
    
      
    
  </body>
</html>