



<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
      
      
        <meta name="author" content="Marcus D. R. Klarqvist">
      
      
        <meta name="lang:clipboard.copy" content="Copy to clipboard">
      
        <meta name="lang:clipboard.copied" content="Copied to clipboard">
      
        <meta name="lang:search.language" content="en">
      
        <meta name="lang:search.pipeline.stopwords" content="True">
      
        <meta name="lang:search.pipeline.trimmer" content="True">
      
        <meta name="lang:search.result.none" content="No matching documents">
      
        <meta name="lang:search.result.one" content="1 matching document">
      
        <meta name="lang:search.result.other" content="# matching documents">
      
        <meta name="lang:search.tokenizer" content="[\s\-]+">
      
      <link rel="shortcut icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.0.4, mkdocs-material-3.2.0">
    
    
      
        <title>CLI - Tomahawk</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/application.572ca0f0.css">
      
      
    
    
      <script src="../assets/javascripts/modernizr.8c900955.js"></script>
    
    
      
        <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700|Roboto+Mono">
        <style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
      
    
    <link rel="stylesheet" href="../assets/fonts/material-icons.css">
    
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
    
  </head>
  
    <body dir="ltr">
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448"
    viewBox="0 0 416 448" id="__github">
  <path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19-18.125
        8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19 18.125-8.5
        18.125 8.5 10.75 19 3.125 20.5zM320 304q0 10-3.125 20.5t-10.75
        19-18.125 8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19
        18.125-8.5 18.125 8.5 10.75 19 3.125 20.5zM360
        304q0-30-17.25-51t-46.75-21q-10.25 0-48.75 5.25-17.75 2.75-39.25
        2.75t-39.25-2.75q-38-5.25-48.75-5.25-29.5 0-46.75 21t-17.25 51q0 22 8
        38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0
        37.25-1.75t35-7.375 30.5-15 20.25-25.75 8-38.375zM416 260q0 51.75-15.25
        82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5-41.75
        1.125q-19.5 0-35.5-0.75t-36.875-3.125-38.125-7.5-34.25-12.875-30.25-20.25-21.5-28.75q-15.5-30.75-15.5-82.75
        0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25
        30.875q36.75-8.75 77.25-8.75 37 0 70 8 26.25-20.5
        46.75-30.25t47.25-9.75q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34
        99.5z" />
</svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
    
      <a href="#getting-started-with-tomahawk" tabindex="1" class="md-skip">
        Skip to content
      </a>
    
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href=".." title="Tomahawk" class="md-header-nav__button md-logo">
          
            <img src="../images/twk_logo_white.svg" height="24">
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          
            
              <span class="md-header-nav__topic">
                Tomahawk
              </span>
              <span class="md-header-nav__topic">
                CLI
              </span>
            
          
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          
            <label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
            
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="__search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
          
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  


  <a href="https://github.com/mklarqvist/tomahawk/" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#__github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
        
      
      
      <main class="md-main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <a href=".." title="Tomahawk" class="md-nav__button md-logo">
      
        <img src="../images/twk_logo_white.svg" width="48" height="48">
      
    </a>
    Tomahawk
  </label>
  
    <div class="md-nav__source">
      


  


  <a href="https://github.com/mklarqvist/tomahawk/" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#__github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href=".." title="Home" class="md-nav__link">
      Home
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../introduction/" title="Introduction" class="md-nav__link">
      Introduction
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../installation/" title="Installation" class="md-nav__link">
      Installation
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-4" type="checkbox" id="nav-4">
    
    <label class="md-nav__link" for="nav-4">
      Technical details
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-4">
        Technical details
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../algorithm-overview/" title="Algorithm overview" class="md-nav__link">
      Algorithm overview
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../aggregation/" title="Aggregation" class="md-nav__link">
      Aggregation
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../job-balancing/" title="Job balancing" class="md-nav__link">
      Job balancing
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-5" type="checkbox" id="nav-5" checked>
    
    <label class="md-nav__link" for="nav-5">
      Tutorial
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-5">
        Tutorial
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    
      
    
    
      <label class="md-nav__link md-nav__link--active" for="__toc">
        CLI
      </label>
    
    <a href="./" title="CLI" class="md-nav__link md-nav__link--active">
      CLI
    </a>
    
      
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#about" title="About" class="md-nav__link">
    About
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#usage-instructions" title="Usage instructions" class="md-nav__link">
    Usage instructions
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#importing-into-tomahawk" title="Importing into Tomahawk" class="md-nav__link">
    Importing into Tomahawk
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#computing-linkage-disequilibrium" title="Computing linkage-disequilibrium" class="md-nav__link">
    Computing linkage-disequilibrium
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#all-pairwise-comparisons" title="All pairwise comparisons" class="md-nav__link">
    All pairwise comparisons
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sliding-window" title="Sliding window" class="md-nav__link">
    Sliding window
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#single-interval-vs-its-local-neighbourhood" title="Single interval vs its local neighbourhood" class="md-nav__link">
    Single interval vs its local neighbourhood
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#concatenating-multiple-archives" title="Concatenating multiple archives" class="md-nav__link">
    Concatenating multiple archives
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sorting-output-files" title="Sorting output files" class="md-nav__link">
    Sorting output files
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#format-descriptions" title="Format descriptions" class="md-nav__link">
    Format descriptions
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ld-format" title="LD format" class="md-nav__link">
    LD format
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#viewing-and-manipulating-output-data" title="Viewing and manipulating output data" class="md-nav__link">
    Viewing and manipulating output data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#aggregating-and-visualizing-datasets" title="Aggregating and visualizing datasets" class="md-nav__link">
    Aggregating and visualizing datasets
  </a>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
    
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../r-tutorial/" title="R" class="md-nav__link">
      R
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../docker-tutorial/" title="Docker" class="md-nav__link">
      Docker
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-6" type="checkbox" id="nav-6">
    
    <label class="md-nav__link" for="nav-6">
      Command line interface
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-6">
        Command line interface
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../cli/cli-aggregate/" title="aggregate" class="md-nav__link">
      aggregate
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../cli/cli-calc/" title="calc" class="md-nav__link">
      calc
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../cli/cli-concat/" title="concat" class="md-nav__link">
      concat
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../cli/cli-import/" title="import" class="md-nav__link">
      import
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../cli/cli-sort/" title="sort" class="md-nav__link">
      sort
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../cli/cli-view/" title="view" class="md-nav__link">
      view
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../cli/cli-haplotype/" title="haplotype" class="md-nav__link">
      haplotype
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../cli/cli-decay/" title="decay" class="md-nav__link">
      decay
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../api-documentation/" title="API documentation" class="md-nav__link">
      API documentation
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../developer-documentation/" title="Developer documentation" class="md-nav__link">
      Developer documentation
    </a>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#about" title="About" class="md-nav__link">
    About
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#usage-instructions" title="Usage instructions" class="md-nav__link">
    Usage instructions
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#importing-into-tomahawk" title="Importing into Tomahawk" class="md-nav__link">
    Importing into Tomahawk
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#computing-linkage-disequilibrium" title="Computing linkage-disequilibrium" class="md-nav__link">
    Computing linkage-disequilibrium
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#all-pairwise-comparisons" title="All pairwise comparisons" class="md-nav__link">
    All pairwise comparisons
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sliding-window" title="Sliding window" class="md-nav__link">
    Sliding window
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#single-interval-vs-its-local-neighbourhood" title="Single interval vs its local neighbourhood" class="md-nav__link">
    Single interval vs its local neighbourhood
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#concatenating-multiple-archives" title="Concatenating multiple archives" class="md-nav__link">
    Concatenating multiple archives
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sorting-output-files" title="Sorting output files" class="md-nav__link">
    Sorting output files
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#format-descriptions" title="Format descriptions" class="md-nav__link">
    Format descriptions
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ld-format" title="LD format" class="md-nav__link">
    LD format
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#viewing-and-manipulating-output-data" title="Viewing and manipulating output data" class="md-nav__link">
    Viewing and manipulating output data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#aggregating-and-visualizing-datasets" title="Aggregating and visualizing datasets" class="md-nav__link">
    Aggregating and visualizing datasets
  </a>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/mklarqvist/tomahawk/edit/master/docs/tutorial.md" title="Edit this page" class="md-icon md-content__icon">&#xE3C9;</a>
                
                
                <h1 id="getting-started-with-tomahawk">Getting started with <code>tomahawk</code><a class="headerlink" href="#getting-started-with-tomahawk" title="Permanent link">&para;</a></h1>
<h2 id="about">About<a class="headerlink" href="#about" title="Permanent link">&para;</a></h2>
<p>This is an introductory tutorial for using Tomahawk. It will cover:</p>
<ul>
<li>Importing data into Tomahawk</li>
<li>Computing linkage-disequilibrium</li>
<li>Subsetting and filtering output data</li>
<li>Aggregating datasets</li>
</ul>
<h2 id="usage-instructions">Usage instructions<a class="headerlink" href="#usage-instructions" title="Permanent link">&para;</a></h2>
<p>The CLI of Tomahawk comprises of several distinct subroutines (listed below).
Executing <code>tomahawk</code> gives a list of commands with brief descriptions and
<code>tomahawk &lt;command&gt;</code> gives detailed details for that command.</p>
<p>All primary Tomahawk commands operate on the binary Tomahawk <code>twk</code> and Tomahawk
output <code>two</code> file format. Interconversions between <code>twk</code> and <code>vcf</code>/<code>bcf</code> is
supported through the commands <code>import</code> for <code>vcf</code>/<code>bcf</code>-&gt;<code>twk</code> and <code>view</code> for
<code>twk</code>-&gt;<code>vcf</code>. Linkage disequilibrium data is written out in binary <code>two</code> format
or human-readable <code>ld</code> format.</p>
<table>
<thead>
<tr>
<th>Command</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="cli/cli-aggregate"><code>aggregate</code></a></td>
<td>data rasterization framework for <code>TWO</code> files</td>
</tr>
<tr>
<td><a href="cli/cli-calc"><code>calc</code></a></td>
<td>calculate linkage disequilibrium</td>
</tr>
<tr>
<td><a href="cli/cli-scalc"><code>scalc</code></a></td>
<td>calculate linkage disequilibrium for a single site</td>
</tr>
<tr>
<td><a href="cli/cli-concat"><code>concat</code></a></td>
<td>concatenate <code>TWO</code> files from the same set of samples</td>
</tr>
<tr>
<td><a href="cli/cli-import"><code>import</code></a></td>
<td>import <code>VCF</code>/<code>VCF.gz</code>/<code>BCF</code> to <code>TWK</code></td>
</tr>
<tr>
<td><a href="cli/cli-sort"><code>sort</code></a></td>
<td>sort <code>TWO</code> file</td>
</tr>
<tr>
<td><a href="cli/cli-view"><code>view</code></a></td>
<td><code>TWO</code>-&gt;<code>LD</code>/<code>TWO</code> view, <code>TWO</code> subset and filter</td>
</tr>
<tr>
<td><a href="cli/cli-haplotype"><code>haplotype</code></a></td>
<td>extract per-sample haplotype strings in <code>FASTA</code>/binary format</td>
</tr>
<tr>
<td><a href="cli/cli-relationship"><code>relationship</code></a></td>
<td>compute marker-based pair-wise sample relationship matrices</td>
</tr>
<tr>
<td><a href="cli/cli-decay"><code>decay</code></a></td>
<td>compute LD-decay over distance</td>
</tr>
<tr>
<td><a href="cli/cli-prune"><code>prune</code></a></td>
<td>perform graph-based LD-pruning of variant sites</td>
</tr>
</tbody>
</table>
<h2 id="importing-into-tomahawk">Importing into Tomahawk<a class="headerlink" href="#importing-into-tomahawk" title="Permanent link">&para;</a></h2>
<p>By design Tomahawk only operates on diploid and bi-allelic SNVs and as such
filters out indels and complex variants. Tomahawk does not support mixed phasing
of genotypes in the same variant (e.g. <code>0|0</code>, <code>0/1</code>). If mixed phasing is found
for a record, all genotypes for that site are converted to unphased genotypes.
This is a conscious design choice as this will internally invoke the correct
algorithm to use for mixed-phase cases.  </p>
<p>Importing standard files to Tomahawk involes using the <code>import</code> command. The
following command imports a <code>bcf</code> file and outputs <code>file.twk</code> while filtering
out variants with &gt;20% missingness and sites that deviate from <a href="https://en.wikipedia.org/wiki/Hardy%E2%80%93Weinberg_principle">Hardy-Weinberg</a>
equilibrium with a probability &lt; 0.001.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>tomahawk import -i file.bcf -o file -m <span class="m">0</span>.2 -H 1e-3
</pre></div>
</td></tr></table>

<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>$ md5sum 1kgp3_chr6.bcf
8c05554ebe1a51e99be2471c96fad4d9  1kgp3_chr6.bcf
</pre></div>
</td></tr></table>

<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>$ bcftools view 1kgp3_chr6.bcf -HG <span class="p">|</span> wc -l
<span class="m">5024119</span>
</pre></div>
</td></tr></table>

<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>$ <span class="nb">time</span> tomahawk import -i 1kgp3_chr6.bcf -o 1kgp3_chr6
</pre></div>
</td></tr></table>

<div class="admonition note">
<p class="admonition-title">Auto-completion of file extensions</p>
<p>You do not have to append the <code>.twk</code> suffix to the output name as Tomahawk
will automatically add this if missing. In the example above <code>"1kgp3_chr6"</code>
will be converted to <code>"1kgp3_chr6.twk"</code> automatically. This is true for 
most Tomahawk commands when using the CLI.</p>
</div>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>Program:   tomahawk-264d039a-dirty (Tools for computing, querying and storing LD data)
Libraries: tomahawk-0.7.0; ZSTD-1.3.8; htslib 1.9
Contact: Marcus D. R. Klarqvist &lt;mk819@cam.ac.uk&gt;
Documentation: https://github.com/mklarqvist/tomahawk
License: MIT
----------
[2019-01-21 11:58:15,692][LOG] Calling import...
[2019-01-21 11:58:15,692][LOG][READER] Opening 1kgp3_chr6.bcf...
[2019-01-21 11:58:15,695][LOG][VCF] Constructing lookup table for 86 contigs...
[2019-01-21 11:58:15,695][LOG][VCF] Samples: 2,504...
[2019-01-21 11:58:15,695][LOG][WRITER] Opening 1kgp3_chr6.twk...
[2019-01-21 11:58:39,307][LOG] Duplicate site dropped: 6:18233985
[2019-01-21 11:59:26,749][LOG] Duplicate site dropped: 6:55137646
[2019-01-21 11:59:39,315][LOG] Duplicate site dropped: 6:67839893
[2019-01-21 11:59:47,176][LOG] Duplicate site dropped: 6:74373442
[2019-01-21 11:59:51,440][LOG] Duplicate site dropped: 6:77843171
[2019-01-21 12:00:41,784][LOG] Duplicate site dropped: 6:121316830
[2019-01-21 12:01:12,830][LOG] Duplicate site dropped: 6:148573620
[2019-01-21 12:01:16,557][LOG] Duplicate site dropped: 6:151397786
<span class="hll">[2019-01-21 12:01:42,644][LOG] Wrote: 4,784,608 variants to 9,570 blocks...
</span>[2019-01-21 12:01:42,644][LOG] Finished: 03m26,952s
<span class="hll">[2019-01-21 12:01:42,644][LOG] Filtered out 239,511 sites (4.76722%):
</span><span class="hll">[2019-01-21 12:01:42,645][LOG]    Invariant: 15,485 (0.308213%)
</span>[2019-01-21 12:01:42,645][LOG]    Missing threshold: 0 (0%)
[2019-01-21 12:01:42,645][LOG]    Insufficient samples: 0 (0%)
[2019-01-21 12:01:42,645][LOG]    Mixed ploidy: 0 (0%)
[2019-01-21 12:01:42,645][LOG]    No genotypes: 0 (0%)
[2019-01-21 12:01:42,645][LOG]    No FORMAT: 0 (0%)
<span class="hll">[2019-01-21 12:01:42,645][LOG]    Not biallelic: 26,277 (0.523017%)
</span><span class="hll">[2019-01-21 12:01:42,645][LOG]    Not SNP: 194,566 (3.87264%)
</span></pre></div>
</td></tr></table>

<h2 id="computing-linkage-disequilibrium">Computing linkage-disequilibrium<a class="headerlink" href="#computing-linkage-disequilibrium" title="Permanent link">&para;</a></h2>
<h3 id="all-pairwise-comparisons">All pairwise comparisons<a class="headerlink" href="#all-pairwise-comparisons" title="Permanent link">&para;</a></h3>
<p>In this first example, we will compute all-vs-all LD associations using the data we imported
in the previous section. To limit compute, we restrict our attention to a 1/45 section of the
data by passing the <code>-c</code> and <code>-C</code> job parameters. We will be using 8 threads (<code>-t</code>), but you
may need to modify this to match the hardware available on your host machine. This job involves
comparing a pair of variants &gt;141 billion times and as such takes around 30 min to finish on
most machines.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>$ tomahawk calc -pi 1kgp3_chr6.twk -o 1kgp3_chr6_1_45 -C <span class="m">1</span> -c <span class="m">45</span> -t <span class="m">8</span>
</pre></div>
</td></tr></table>

<div class="admonition note">
<p class="admonition-title">Valid job balancing partitions</p>
<p>When computing genome-wide LD the
balancing requires that number of sub-problems (<code>-c</code>) must be a member of the function: 
$$
\binom{c}{2} + c = \frac{c^2 + c}{2}, c &gt; 0
$$
This function is equivalent to the upper-triangular of a square
(<code>c</code>-by-<code>c</code>) matrix plus the diagonal. Read more about <a href="../job-balancing/">load
partitioning</a> in Tomahawk.</p>
<p>Here is the first 100 valid partition sizes:<br />
<code>1</code>, <code>3</code>, <code>6</code>, <code>10</code>, <code>15</code>, <code>21</code>, <code>28</code>, <code>36</code>, <code>45</code>, <code>55</code>, <code>66</code>, <code>78</code>, <code>91</code>, <code>105</code>, <code>120</code>, <code>136</code>, <code>153</code>, <code>171</code>, <code>190</code>, <code>210</code>, <code>231</code>, <code>253</code>, <code>276</code>, <code>300</code>, <code>325</code>, <code>351</code>, <code>378</code>, <code>406</code>, <code>435</code>, <code>465</code>, <code>496</code>, <code>528</code>, <code>561</code>, <code>595</code>, <code>630</code>, <code>666</code>, <code>703</code>, <code>741</code>, <code>780</code>, <code>820</code>, <code>861</code>, <code>903</code>, <code>946</code>, <code>990</code>, <code>1035</code>, <code>1081</code>, <code>1128</code>, <code>1176</code>, <code>1225</code>, <code>1275</code>, <code>1326</code>, <code>1378</code>, <code>1431</code>, <code>1485</code>, <code>1540</code>, <code>1596</code>, <code>1653</code>, <code>1711</code>, <code>1770</code>, <code>1830</code>, <code>1891</code>, <code>1953</code>, <code>2016</code>, <code>2080</code>, <code>2145</code>, <code>2211</code>, <code>2278</code>, <code>2346</code>, <code>2415</code>, <code>2485</code>, <code>2556</code>, <code>2628</code>, <code>2701</code>, <code>2775</code>, <code>2850</code>, <code>2926</code>, <code>3003</code>, <code>3081</code>, <code>3160</code>, <code>3240</code>, <code>3321</code>, <code>3403</code>, <code>3486</code>, <code>3570</code>, <code>3655</code>, <code>3741</code>, <code>3828</code>, <code>3916</code>, <code>4005</code>, <code>4095</code>, <code>4186</code>, <code>4278</code>, <code>4371</code>, <code>4465</code>, <code>4560</code>, <code>4656</code>, <code>4753</code>, <code>4851</code>, <code>4950</code>, <code>5050</code></p>
</div>
<div class="admonition warning">
<p class="admonition-title">Massive workload</p>
<p>By default, Tomahawk will attempt to compute the genome-wide linkage-disequilibrium
for all the provided variants and samples. This generally require huge amount
of pairwise variant comparisons. Depending on your interests, you may not want to
undertake this large compute and need to parameterize for your particular use-case.</p>
</div>
<div class="admonition hint">
<p class="admonition-title">Different load balancing</p>
<p>As a consequence of the job balancing approach in Tomahawk, the diagonal jobs
will always involve less variant comparisons (\( \binom{n}{2} \)) compared to off-diagonal
jobs (\(n \times m\)), where \(n\) and \(m\) are the number of records in either block. 
Keep this in mind if you are choosing a partition size given
the run-times of a given job. For example, examine the total run-time of the second
job (<code>-C 2</code>) &mdash; the first off-diagonal job &mdash; as a proxy for the expected average run-time per job.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Output ordering</p>
<p>By design, Tomahawk computes pairwise associations using an <a href="https://simple.wikipedia.org/wiki/Out-of-order_execution">out-of-order execution</a> 
paradigm and further permutes the input ordering by using <a href="https://en.wikipedia.org/wiki/Loop_nest_optimization">cache-blocking</a>.
Additionally, each thread/slave transiently store a number of computed associations in a
private buffer that will be flushed upon reaching some frequency threshold. Because of these
technicalities, Tomahawk will neither produce ordered output nor will the permutation of the
output order be identical between runs on identical data. This has no practical consequence
to most downstream applications in Tomahawk with the exception of query speeds. To address
this limitation we have introduced a powerful sorting paradigm that will be discussed below.</p>
</div>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>Program:   tomahawk-264d039a-dirty (Tools for computing, querying and storing LD data)
Libraries: tomahawk-0.7.0; ZSTD-1.3.8; htslib 1.9
Contact: Marcus D. R. Klarqvist &lt;mk819@cam.ac.uk&gt;
Documentation: https://github.com/mklarqvist/tomahawk
License: MIT
----------
[2019-01-21 12:13:50,210][LOG] Calling calc...
[2019-01-21 12:13:50,211][LOG][READER] Opening 1kgp3_chr6.twk...
<span class="hll">[2019-01-21 12:13:50,212][LOG] Samples: 2,504...
</span><span class="hll">[2019-01-21 12:13:50,212][LOG][BALANCING] Using ranges [0-1063,0-1063] in square mode...
</span>[2019-01-21 12:13:50,212][LOG] Allocating 1,063 blocks...
[2019-01-21 12:13:50,213][LOG] Running in standard mode. Pre-computing data...
[2019-01-21 12:13:50,213][LOG][SIMD] Vectorized instructions available: SSE4...
[2019-01-21 12:13:50,213][LOG] Constructing list, vector, RLE...
[2019-01-21 12:13:50,213][LOG][THREAD] Unpacking using 7 threads: ....... Done! 01,291s
<span class="hll">[2019-01-21 12:13:51,504][LOG] 531,500 variants from 1,063 blocks...
</span>[2019-01-21 12:13:51,504][LOG][PARAMS] square=TRUE,window=FALSE,low_memory=FALSE,bitmaps=FALSE,single=FALSE,force_phased=TRUE,force_unphased=FALSE,compression_level=1,block_size=500,output_block_size=10000,l_surrounding=500000,minP=1.000000,minR2=0.100000,maxR2=100.000000,minDprime=0.000000,maxDprime=100.000000,n_chunks=45,c_chunk=0,n_threads=8,ldd_type=3,cycle_threshold=0
<span class="hll">[2019-01-21 12:13:51,504][LOG] Performing: 141,245,859,250 variant comparisons...
</span>[2019-01-21 12:13:51,504][LOG][WRITER] Opening 1kgp3_chr6_1_45.two...
[2019-01-21 12:13:51,505][LOG][THREAD] Spawning 8 threads: ........
[2019-01-21 12:13:51,533][PROGRESS] Time elapsed       Variants           Genotypes         Output  Progress    Est. Time left
[2019-01-21 12:14:21,533][PROGRESS]      30,000s  3,685,746,500   9,229,109,236,000        988,858   2.60945%   18m39s
[2019-01-21 12:14:51,533][PROGRESS]   01m00,000s  6,425,868,750  16,090,375,350,000      1,689,422   4.54942%   20m58s
[truncated]
[2019-01-21 12:39:51,546][PROGRESS]   26m00,013s   140,131,382,750 350,888,982,406,000     48,635,114    99.211%    12s
[2019-01-21 12:40:04,317][PROGRESS] Finished in 26m12,784s. Variants: 141,245,859,250, genotypes: 353,679,631,562,000, output: 49,870,388
[2019-01-21 12:40:04,317][PROGRESS] 89,806,242 variants/s and 224,874,830,855 genotypes/s
[2019-01-21 12:40:04,321][LOG][PROGRESS] All done...26m12,815s!
</pre></div>
</td></tr></table>

<p>This run generated 935817820 bytes (935.8 MB) of output data in binary format using the default
compression level (1):</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>$ ls -l 1kgp3_chr6_1_45.two
-rw-rw-r-- <span class="m">1</span> mk819 mk819 <span class="m">935817820</span> Jan <span class="m">21</span> <span class="m">12</span>:40 1kgp3_chr6_1_45.two
</pre></div>
</td></tr></table>

<p>If this data was stored in plain, human-readable, text format it would use 
over 45 GB:</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>$ tomahawk view -i 1kgp3_chr6_1_45.two <span class="p">|</span> wc -c
<span class="m">4544702179</span>
</pre></div>
</td></tr></table>

<h3 id="sliding-window">Sliding window<a class="headerlink" href="#sliding-window" title="Permanent link">&para;</a></h3>
<p>If you are working with a species with well-know LD structure (such as humans) you can reduce the computational cost by limiting the search-space to a fixed-sized sliding window (<code>-w</code>). In window mode you are free to choose any arbitrary sub-problem (<code>-c</code>) size.</p>
<div class="admonition note">
<p class="admonition-title">Window properties</p>
<p>By default, Tomahawk computes the linkage disequilibrium associations for all pairs
of variants within <code>-w</code> bases of the target SNV on either direction. In order to
maximize computational throughput, we made the design decision to <em>maintain</em> blocks
of data that overlap to the target interval. This has the consequence that adjacent
data that may <strong>not</strong> directly overlap with the target region will still be computed. The 
technical reasoning for this is, in simplified terms, that the online repackaging of internal data
blocks is more expensive than computing a small (relatively) number of non-desired (off-target) associations.</p>
</div>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="nb">time</span> tomahawk calc -pi 1kgp3_chr6.twk -o 1kgp3_chr6_4mb -w <span class="m">4000000</span>
</pre></div>
</td></tr></table>

<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>Program:   tomahawk-264d039a-dirty (Tools for computing, querying and storing LD data)
Libraries: tomahawk-0.7.0; ZSTD-1.3.8; htslib 1.9
Contact: Marcus D. R. Klarqvist &lt;mk819@cam.ac.uk&gt;
Documentation: https://github.com/mklarqvist/tomahawk
License: MIT
----------
[2019-01-21 14:10:33,616][LOG] Calling calc...
[2019-01-21 14:10:33,616][LOG][READER] Opening 1kgp3_chr6.twk...
[2019-01-21 14:10:33,619][LOG] Samples: 2,504...
[2019-01-21 14:10:33,619][LOG][BALANCING] Using ranges [0-9570,0-9570] in window mode...
[2019-01-21 14:10:33,619][LOG] Allocating 9,570 blocks...
[2019-01-21 14:10:33,620][LOG] Running in standard mode. Pre-computing data...
[2019-01-21 14:10:33,620][LOG][SIMD] Vectorized instructions available: SSE4...
[2019-01-21 14:10:33,620][LOG] Constructing list, vector, RLE...
[2019-01-21 14:10:33,620][LOG][THREAD] Unpacking using 7 threads: ....... Done! 15,774s
[2019-01-21 14:10:49,394][LOG] 4,784,608 variants from 9,570 blocks...
[2019-01-21 14:10:49,394][LOG][PARAMS] square=TRUE,window=TRUE,low_memory=FALSE,bitmaps=FALSE,single=FALSE,force_phased=TRUE,force_unphased=FALSE,compression_level=1,block_size=500,output_block_size=10000,window_size=4000000,l_surrounding=500000,minP=1.000000,minR2=0.100000,maxR2=100.000000,minDprime=0.000000,maxDprime=100.000000,n_chunks=1,c_chunk=0,n_threads=8,ldd_type=3,cycle_threshold=0
[2019-01-21 14:10:49,394][LOG] Performing: 11,446,234,464,528 variant comparisons...
[2019-01-21 14:10:49,394][LOG][WRITER] Opening 1kgp3_chr6_4mb.two...
[2019-01-21 14:10:49,402][LOG][THREAD] Spawning 8 threads: ........
[2019-01-21 14:10:49,424][PROGRESS] Time elapsed       Variants           Genotypes         Output  Progress    Est. Time left
[2019-01-21 14:11:19,424][PROGRESS]      30,000s  2,311,240,500   5,787,346,212,000        853,368         0    0
[2019-01-21 14:11:49,424][PROGRESS]   01m00,000s  4,562,731,500  11,425,079,676,000      1,717,956         0    0
[truncated]
[2019-01-21 16:12:49,492][PROGRESS] 02h02m00,068s   527,889,726,500  1,321,835,875,156,000    470,354,846         0 0
[2019-01-21 16:13:17,236][PROGRESS] Finished in 02h02m27,812s. Variants: 529,807,522,528, genotypes: 1,326,638,036,410,112, output: 473,514,018
[2019-01-21 16:13:17,237][PROGRESS] 72,104,114 variants/s and 180,548,701,880 genotypes/s
[2019-01-21 16:13:17,257][LOG][PROGRESS] All done...02h02m27,854s!
</pre></div>
</td></tr></table>

<h3 id="single-interval-vs-its-local-neighbourhood">Single interval vs its local neighbourhood<a class="headerlink" href="#single-interval-vs-its-local-neighbourhood" title="Permanent link">&para;</a></h3>
<p>In many cases we are not interested in computing large-scale
linkage-disequilibrium associations but have a limited genomics window of
interest. If this is the case, you can use the specialization subroutine of
<code>calc</code> called <code>scalc</code> that is parameterizable with a target region and a
neighbourhood size in bases.</p>
<p>In this example, we will compute the regional associations of variants mapping
to the interval chr6:10e6-10.1e6 and a 100kb neighbourhood.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>tomahawk scalc -i 1kgp3_chr6.twk -o <span class="nb">test</span> -I <span class="m">6</span>:10e6-10.1e6 -w <span class="m">100000</span>
</pre></div>
</td></tr></table>

<div class="admonition note">
<p class="admonition-title">Neighbourhood interval</p>
<p>The neighbourhood is computed as the region from [start of interval -
neighbourhood, start of interval) and (end of interval + neighbourhood). If the
start of the interval minus the neighbourhood falls outside the chromosome
(position &lt; 0) then this interval is truncated to 0 in the left end.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Scalability</p>
<p>This subroutine was designed to be used with a relatively short interval to
maximize computability in this case. It is however possible to provide any valid
interval as the target region to compute linkage-disequilibrium for. Providing
large intervals will result in poor performance and potentially undesired
output.</p>
</div>
<h2 id="concatenating-multiple-archives">Concatenating multiple archives<a class="headerlink" href="#concatenating-multiple-archives" title="Permanent link">&para;</a></h2>
<p>On of the immediate downsides of partitioning compute into multiple
non-overlapping sub-problems is that we will generate a large number of
independent files that must be merged prior to downstream analysis. Fortunatly,
this is a trivial operation in Tomahawk and involves the <code>concat</code> command.</p>
<p>First lets compute some data using the first 3 out of 990 partitions of the
dataset used above.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="k">for</span> i in <span class="o">{</span><span class="m">1</span>..3<span class="o">}</span><span class="p">;</span> <span class="k">do</span> <span class="nb">time</span> tomahawk calc -pi 1kgp3_chr6.twk -c <span class="m">990</span> -C <span class="nv">$i</span> -o part<span class="nv">$i</span><span class="se">\_</span><span class="m">3</span>.two<span class="p">;</span> <span class="k">done</span>
</pre></div>
</td></tr></table>

<p>Next, since we only have three files, we can concatenate (merge) these files together
into a single archieve by passing each file name to the command:</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>$ <span class="nb">time</span> tomahawk concat -i part1_3.two -i part2_3.two -i part3_3.two -o part3_concat
<span class="o">[</span><span class="m">2019</span>-01-22 <span class="m">10</span>:55:47,799<span class="o">][</span>LOG<span class="o">]</span> All files are compatible. Beginning merging...
<span class="o">[</span><span class="m">2019</span>-01-22 <span class="m">10</span>:55:47,800<span class="o">][</span>LOG<span class="o">]</span> Appending part1_3.two... <span class="m">397</span>.953344 Mb/98.987848 Mb
<span class="o">[</span><span class="m">2019</span>-01-22 <span class="m">10</span>:55:47,889<span class="o">][</span>LOG<span class="o">]</span> Appending part2_3.two... <span class="m">359</span>.538884 Mb/54.763322 Mb
<span class="o">[</span><span class="m">2019</span>-01-22 <span class="m">10</span>:55:47,938<span class="o">][</span>LOG<span class="o">]</span> Appending part3_3.two... <span class="m">346</span>.075500 Mb/52.072741 Mb
<span class="o">[</span><span class="m">2019</span>-01-22 <span class="m">10</span>:55:47,988<span class="o">][</span>LOG<span class="o">]</span> Finished. Added <span class="m">3</span> files...
<span class="o">[</span><span class="m">2019</span>-01-22 <span class="m">10</span>:55:47,988<span class="o">][</span>LOG<span class="o">]</span> Total size: <span class="nv">Uncompressed</span> <span class="o">=</span> <span class="m">1</span>.103568 Gb and <span class="nv">compressed</span> <span class="o">=</span> <span class="m">205</span>.823911 Mb

real    0m0.226s
user    0m0.036s
sys     0m0.190s
</pre></div>
</td></tr></table>

<p>Passing every single input file in the command line does not become feasble when we have
many hundreds to thousands of files. To address this, it is possible to first store the
target file paths in a text file and then pass that to Tomahawk. Using the data from above,
we save these file paths to the file <code>part_file_list.txt</code>.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="k">for</span> i in <span class="o">{</span><span class="m">1</span>..3<span class="o">}</span><span class="p">;</span> <span class="k">do</span> <span class="nb">echo</span> part<span class="nv">$i</span><span class="se">\_</span><span class="m">3</span>.two &gt;&gt; part_file_list.txt<span class="p">;</span> <span class="k">done</span>
</pre></div>
</td></tr></table>

<p>Then we simply pass this list of file paths to Tomahawk using the <code>-I</code> argument:</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>$ <span class="nb">time</span> tomahawk concat -I part_file_list.txt -o part3_concat 
adding <span class="nv">file</span><span class="o">=</span>part1_3.two
adding <span class="nv">file</span><span class="o">=</span>part2_3.two
adding <span class="nv">file</span><span class="o">=</span>part3_3.two
<span class="o">[</span><span class="m">2019</span>-01-22 <span class="m">10</span>:57:43,227<span class="o">][</span>LOG<span class="o">]</span> All files are compatible. Beginning merging...
<span class="o">[</span><span class="m">2019</span>-01-22 <span class="m">10</span>:57:43,262<span class="o">][</span>LOG<span class="o">]</span> Appending part1_3.two... <span class="m">397</span>.953344 Mb/98.987848 Mb
<span class="o">[</span><span class="m">2019</span>-01-22 <span class="m">10</span>:57:43,350<span class="o">][</span>LOG<span class="o">]</span> Appending part2_3.two... <span class="m">359</span>.538884 Mb/54.763322 Mb
<span class="o">[</span><span class="m">2019</span>-01-22 <span class="m">10</span>:57:43,400<span class="o">][</span>LOG<span class="o">]</span> Appending part3_3.two... <span class="m">346</span>.075500 Mb/52.072741 Mb
<span class="o">[</span><span class="m">2019</span>-01-22 <span class="m">10</span>:57:43,447<span class="o">][</span>LOG<span class="o">]</span> Finished. Added <span class="m">3</span> files...
<span class="o">[</span><span class="m">2019</span>-01-22 <span class="m">10</span>:57:43,447<span class="o">][</span>LOG<span class="o">]</span> Total size: <span class="nv">Uncompressed</span> <span class="o">=</span> <span class="m">1</span>.103568 Gb and <span class="nv">compressed</span> <span class="o">=</span> <span class="m">205</span>.823911 Mb

real    0m0.477s
user    0m0.040s
sys     0m0.241s
</pre></div>
</td></tr></table>

<div class="admonition danger">
<p class="admonition-title">Possible duplications</p>
<p>The <code>concat</code> subroutine will dedupe input file paths but will <strong>neiter check nor
guarantee</strong> that the actual input data is not duplicated. Therefore, it is possible to get
duplicates in your data if you are not careful. These duplicate entries could corrupt
any downstream insights!</p>
</div>
<h2 id="sorting-output-files">Sorting output files<a class="headerlink" href="#sorting-output-files" title="Permanent link">&para;</a></h2>
<p>All subroutines in Tomahawk will work on unsorted output files. However, sorted
files are <em>much</em> faster to query and will result in considerable time savings.
This is especially true for very large files. Sorting files in Tomahawk is
trivial and involves calling the <code>sort</code> command with the required fields input
(<code>-i</code>) and output (<code>-o</code>).</p>
<div class="admonition warning">
<p class="admonition-title">Disk usage</p>
<p>Tomahawk can easily sort huge <code>.two</code> files with the help of <a href="https://en.wikipedia.org/wiki/External_sorting">external merge
sorting</a> algorithms followed by
a single <a href="https://en.wikipedia.org/wiki/K-way_merge_algorithm">k-way merge</a>,
memory assisted operation. The external merge step require additional disk space
approximately equal to the size of the input dataset. Then, the merge operation
will write a, generally, smaller output file. On average, we can approxiate that
the sort operation require an additional two times the input file size on disk.
After the sorting procedure is completed, the intermediate files are removed.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Temporary files</p>
<p>By default, Tomahawk will generate temporary files in the same directory as the
input file. If this file path is undesired then you have to modify the appropriate
parameters.</p>
</div>
<p>In this example we will sort the almost 500 million (473,514,826) records
computed from the sliding window example above. This archive corresponds to &gt;50
Gb of binary data.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>$ <span class="nb">time</span> tomahawk sort -i 1kgp3_chr6_4mb.two -o 1kgp3_chr6_4mb_sorted
</pre></div>
</td></tr></table>

<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>Program:   tomahawk-264d039a-dirty (Tools for computing, querying and storing LD data)
Libraries: tomahawk-0.7.0; ZSTD-1.3.8; htslib 1.9
Contact: Marcus D. R. Klarqvist &lt;mk819@cam.ac.uk&gt;
Documentation: https://github.com/mklarqvist/tomahawk
License: MIT
----------
[2019-01-22 11:20:43,978][LOG] Calling sort...
[2019-01-22 11:20:43,993][LOG] Blocks: 47,358
[2019-01-22 11:20:43,993][LOG] Uncompressed size: 50.192950 Gb
[2019-01-22 11:20:43,993][LOG] Sorting 473,514,826 records...
[2019-01-22 11:20:43,993][LOG][THREAD] Data/thread: 6.274119 Gb
[2019-01-22 11:20:43,993][PROGRESS] Time elapsed       Variants  Progress   Est. Time left
[2019-01-22 11:20:43,994][LOG][THREAD] Slave-0: range=0-&gt;5919/47358 and name 1kgp3_chr6_4mb_sorted_fileSucRkC.two
[2019-01-22 11:20:43,994][LOG][THREAD] Slave-1: range=5919-&gt;11838/47358 and name 1kgp3_chr6_4mb_sorted_filewPI10T.two
[2019-01-22 11:20:43,994][LOG][THREAD] Slave-2: range=11838-&gt;17757/47358 and name 1kgp3_chr6_4mb_sorted_fileo1zcHb.two
[2019-01-22 11:20:43,994][LOG][THREAD] Slave-3: range=17757-&gt;23676/47358 and name 1kgp3_chr6_4mb_sorted_fileBvRnnt.two
[2019-01-22 11:20:43,994][LOG][THREAD] Slave-4: range=23676-&gt;29595/47358 and name 1kgp3_chr6_4mb_sorted_filewbBz3K.two
[2019-01-22 11:20:43,994][LOG][THREAD] Slave-5: range=29595-&gt;35514/47358 and name 1kgp3_chr6_4mb_sorted_fileCrNLJ2.two
[2019-01-22 11:20:43,994][LOG][THREAD] Slave-6: range=35514-&gt;41433/47358 and name 1kgp3_chr6_4mb_sorted_filerCfYpk.two
[2019-01-22 11:20:43,994][LOG][THREAD] Slave-7: range=41433-&gt;47358/47358 and name 1kgp3_chr6_4mb_sorted_filexx9a6B.two
[2019-01-22 11:21:13,994][PROGRESS]      30,000s     70,594,275   14.9086%  2m51s
[2019-01-22 11:21:43,994][PROGRESS]   01m00,000s    146,732,835    30.988%  2m13s
[2019-01-22 11:22:13,994][PROGRESS]   01m30,000s    245,535,675   51.8539%  1m23s
[2019-01-22 11:22:43,995][PROGRESS]   02m00,001s    323,064,235   68.2268%  55s
[2019-01-22 11:23:13,995][PROGRESS]   02m30,001s    421,592,790   89.0348%  18s
[2019-01-22 11:23:24,951][LOG][THREAD] Finished: 1kgp3_chr6_4mb_sorted_filewbBz3K.two with 23676-29595. Sorted n=59190000 variants with size=1.176156 Gb
[2019-01-22 11:23:25,114][LOG][THREAD] Finished: 1kgp3_chr6_4mb_sorted_fileBvRnnt.two with 17757-23676. Sorted n=59190000 variants with size=1.286938 Gb
[2019-01-22 11:23:25,385][LOG][THREAD] Finished: 1kgp3_chr6_4mb_sorted_filerCfYpk.two with 35514-41433. Sorted n=59190000 variants with size=1.219126 Gb
[2019-01-22 11:23:25,710][LOG][THREAD] Finished: 1kgp3_chr6_4mb_sorted_fileCrNLJ2.two with 29595-35514. Sorted n=59190000 variants with size=1.161081 Gb
[2019-01-22 11:23:26,384][LOG][THREAD] Finished: 1kgp3_chr6_4mb_sorted_filexx9a6B.two with 41433-47358. Sorted n=59184826 variants with size=1.215231 Gb
[2019-01-22 11:23:26,425][LOG][THREAD] Finished: 1kgp3_chr6_4mb_sorted_fileSucRkC.two with 0-5919. Sorted n=59190000 variants with size=1.242923 Gb
[2019-01-22 11:23:29,966][LOG][THREAD] Finished: 1kgp3_chr6_4mb_sorted_fileo1zcHb.two with 11838-17757. Sorted n=59190000 variants with size=1.686758 Gb
[2019-01-22 11:23:31,300][LOG][THREAD] Finished: 1kgp3_chr6_4mb_sorted_filewPI10T.two with 5919-11838. Sorted n=59190000 variants with size=2.013554 Gb
[2019-01-22 11:23:31,338][PROGRESS] 02m47,344s  473,514,826 (2,829,587 variants/s)
[2019-01-22 11:23:31,338][PROGRESS] Finished!
[2019-01-22 11:23:31,341][LOG] Spawning 112 queues with 2.380952 Mb each...
[2019-01-22 11:23:35,090][LOG][WRITER] Opening &quot;1kgp3_chr6_4mb_sorted.two&quot;...
[2019-01-22 11:23:35,091][PROGRESS] Time elapsed       Variants  Progress   Est. Time left
[2019-01-22 11:24:05,091][PROGRESS]      30,000s     39,757,023   8.39615%  5m27s
[2019-01-22 11:24:35,091][PROGRESS]   01m00,000s     76,990,000   16.2593%  5m9s
[2019-01-22 11:25:05,091][PROGRESS]   01m30,000s    107,840,000   22.7744%  5m5s
[2019-01-22 11:25:35,091][PROGRESS]   02m00,000s    133,420,000   28.1765%  5m5s
[2019-01-22 11:26:05,092][PROGRESS]   02m30,000s    170,215,235   35.9472%  4m27s
[2019-01-22 11:26:35,092][PROGRESS]   03m00,001s    208,070,000   43.9416%  3m49s
[2019-01-22 11:27:05,092][PROGRESS]   03m30,001s    247,010,000   52.1652%  3m12s
[2019-01-22 11:27:35,092][PROGRESS]   04m00,001s    287,690,000   60.7563%  2m35s
[2019-01-22 11:28:05,092][PROGRESS]   04m30,001s    325,480,000    68.737%  2m2s
[2019-01-22 11:28:35,093][PROGRESS]   05m00,001s    364,299,966   76.9353%  1m29s
[2019-01-22 11:29:05,093][PROGRESS]   05m30,002s    401,367,427   84.7634%  59s
[2019-01-22 11:29:35,093][PROGRESS]   06m00,002s    439,290,000   92.7722%  28s
[2019-01-22 11:30:02,180][PROGRESS] 06m27,089s  473,514,826 (1,223,269 variants/s)
[2019-01-22 11:30:02,180][PROGRESS] Finished!
[2019-01-22 11:30:02,194][LOG] Finished merging! Time: 06m27,103s
[2019-01-22 11:30:02,194][LOG] Deleting temp files...
[2019-01-22 11:30:02,194][LOG] Deleted 1kgp3_chr6_4mb_sorted_fileSucRkC.two
[2019-01-22 11:30:02,194][LOG] Deleted 1kgp3_chr6_4mb_sorted_filewPI10T.two
[2019-01-22 11:30:02,194][LOG] Deleted 1kgp3_chr6_4mb_sorted_fileo1zcHb.two
[2019-01-22 11:30:02,194][LOG] Deleted 1kgp3_chr6_4mb_sorted_fileBvRnnt.two
[2019-01-22 11:30:02,194][LOG] Deleted 1kgp3_chr6_4mb_sorted_filewbBz3K.two
[2019-01-22 11:30:02,194][LOG] Deleted 1kgp3_chr6_4mb_sorted_fileCrNLJ2.two
[2019-01-22 11:30:02,194][LOG] Deleted 1kgp3_chr6_4mb_sorted_filerCfYpk.two
[2019-01-22 11:30:02,194][LOG] Deleted 1kgp3_chr6_4mb_sorted_filexx9a6B.two
[2019-01-22 11:30:02,678][LOG] Finished!

real    9m18.793s
user    23m31.390s
sys     0m53.197s
</pre></div>
</td></tr></table>

<div class="admonition success">
<p class="admonition-title">Faster queries</p>
<p>Success: You can now use the newly created sorted archieve exactly as unsorted
files but with faster query times.</p>
</div>
<h2 id="format-descriptions">Format descriptions<a class="headerlink" href="#format-descriptions" title="Permanent link">&para;</a></h2>
<div class="admonition note">
<p class="admonition-title">Memory layout (technical)</p>
<p>A frequent design choice when designing a file format is how to align memory of records:
either as lists of records (array-of-struct) versus column-orientated layouts (struct-of-array).
The pivoted layout (struct-of-array) of <code>twk1_two_t</code> structs in Tomahawk will result <em>considerable</em>
savings in disk space usage at the expense of querying speed of the resulting data. We decided to
build the <code>two</code> format around the classical array-of-struct memory layout as we want to maximize
the computability of the data. This is especially true in our application as we will never
support individual columnar slicing and subset operations.</p>
</div>
<h3 id="ld-format"><code>LD</code> format<a class="headerlink" href="#ld-format" title="Permanent link">&para;</a></h3>
<p>Tomahawk can output binary <code>two</code> data in the human-readable <code>ld</code> format by invoking the <code>view</code> command. 
The general schema for <code>ld</code> is below:</p>
<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>FLAG</code></td>
<td>Bit-packed boolean flags (see below)</td>
</tr>
<tr>
<td><code>CHROM_A</code></td>
<td>Chromosome for marker A</td>
</tr>
<tr>
<td><code>POS_A</code></td>
<td>Position for marker A</td>
</tr>
<tr>
<td><code>CHROM_B</code></td>
<td>Chromosome for marker B</td>
</tr>
<tr>
<td><code>POS_B</code></td>
<td>Position for marker B</td>
</tr>
<tr>
<td><code>REF_REF</code></td>
<td>Inner product of (0,0) haplotypes</td>
</tr>
<tr>
<td><code>REF_ALT</code></td>
<td>Inner product of (0,1) haplotypes</td>
</tr>
<tr>
<td><code>ALT_REF</code></td>
<td>Inner product of (1,0) haplotypes</td>
</tr>
<tr>
<td><code>ALT_ALT</code></td>
<td>Inner proruct of (1,1) haplotypes</td>
</tr>
<tr>
<td><a href="https://en.wikipedia.org/wiki/Linkage_disequilibrium"><code>D</code></a></td>
<td>Coefficient of linkage disequilibrium</td>
</tr>
<tr>
<td><code>DPrime</code></td>
<td>Normalized coefficient of linkage disequilibrium (scaled to [-1,1])</td>
</tr>
<tr>
<td><a href="https://en.wikipedia.org/wiki/Pearson_correlation_coefficient"><code>R</code></a></td>
<td>Pearson correlation coefficient</td>
</tr>
<tr>
<td><code>R2</code></td>
<td>Squared pearson correlation coefficient</td>
</tr>
<tr>
<td><a href="https://en.wikipedia.org/wiki/Fisher's_exact_test"><code>P</code></a></td>
<td>Fisher's exact test P-value of the 2x2 haplotype contigency table</td>
</tr>
<tr>
<td><code>ChiSqModel</code></td>
<td>Chi-squared critical value of the 3x3 unphased table of the selected cubic root (&alpha;, &beta;, or &delta;)</td>
</tr>
<tr>
<td><code>ChiSqTable</code></td>
<td>Chi-squared critical value of table (useful comparator when <code>P</code> = 0)</td>
</tr>
</tbody>
</table>
<p>The 2x2 contingency table, or matrix, for the Fisher's exact test (<code>P</code>) for haplotypes look like this:</p>
<table>
<thead>
<tr>
<th></th>
<th>REF-A</th>
<th>REF-B</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>REF-B</strong></td>
<td>A</td>
<td>B</td>
</tr>
<tr>
<td><strong>ALT-B</strong></td>
<td>C</td>
<td>D</td>
</tr>
</tbody>
</table>
<p>The 3x3 contigency table, or matrix, for the Chi-squared test for the unphased model looks like this:</p>
<table>
<thead>
<tr>
<th></th>
<th>0/0</th>
<th>0/1</th>
<th>1/1</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>0/0</strong></td>
<td>A</td>
<td>B</td>
<td>C</td>
</tr>
<tr>
<td><strong>0/1</strong></td>
<td>D</td>
<td>E</td>
<td>F</td>
</tr>
<tr>
<td><strong>1/1</strong></td>
<td>G</td>
<td>H</td>
<td>J</td>
</tr>
</tbody>
</table>
<p>The <code>two</code> <code>FLAG</code> values are bit-packed booleans in a single integer field and describe a variety of states a pair of markers can be in.</p>
<table>
<thead>
<tr>
<th>Bit position</th>
<th>Numeric value</th>
<th>One-hot</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>1</td>
<td>1</td>
<td>000000000000000<strong>1</strong></td>
<td>Used phased math.</td>
</tr>
<tr>
<td>2</td>
<td>2</td>
<td>00000000000000<strong>1</strong>0</td>
<td>Acceptor and donor variants are on the same contig.</td>
</tr>
<tr>
<td>3</td>
<td>4</td>
<td>0000000000000<strong>1</strong>00</td>
<td>Acceptor and donor variants are far apart on the same contig.</td>
</tr>
<tr>
<td>4</td>
<td>8</td>
<td>000000000000<strong>1</strong>000</td>
<td>The output contingency matrix has at least one empty cell (referred to as complete).</td>
</tr>
<tr>
<td>5</td>
<td>16</td>
<td>00000000000<strong>1</strong>0000</td>
<td>Output correlation coefficient is perfect (1.0).</td>
</tr>
<tr>
<td>6</td>
<td>32</td>
<td>0000000000<strong>1</strong>00000</td>
<td>Output solution is one of &gt;1 possible solutions. This only occurs for unphased pairs.</td>
</tr>
<tr>
<td>7</td>
<td>64</td>
<td>000000000<strong>1</strong>000000</td>
<td>Output data was generated in 'fast mode'.</td>
</tr>
<tr>
<td>8</td>
<td>128</td>
<td>00000000<strong>1</strong>0000000</td>
<td>Output data is estimated from a subsampling of the total pool of genotypes.</td>
</tr>
<tr>
<td>9</td>
<td>256</td>
<td>0000000<strong>1</strong>00000000</td>
<td>Donor vector has missing value(s).</td>
</tr>
<tr>
<td>10</td>
<td>512</td>
<td>000000<strong>1</strong>000000000</td>
<td>Acceptor vector has missing value(s).</td>
</tr>
<tr>
<td>11</td>
<td>1024</td>
<td>00000<strong>1</strong>0000000000</td>
<td>Donor vector has low allele count (&lt;5).</td>
</tr>
<tr>
<td>12</td>
<td>2048</td>
<td>0000<strong>1</strong>00000000000</td>
<td>Acceptor vector has low allele count (&lt;5).</td>
</tr>
<tr>
<td>13</td>
<td>4096</td>
<td>000<strong>1</strong>000000000000</td>
<td>Acceptor vector has a HWE-P value &lt; 1e-4.</td>
</tr>
<tr>
<td>14</td>
<td>8192</td>
<td>00<strong>1</strong>0000000000000</td>
<td>Donor vector has a HWE-P value &lt; 1e-4.</td>
</tr>
</tbody>
</table>
<h2 id="viewing-and-manipulating-output-data">Viewing and manipulating output data<a class="headerlink" href="#viewing-and-manipulating-output-data" title="Permanent link">&para;</a></h2>
<p>It is possible to filter <code>two</code> output data by: 
1) either start or end contig e.g. <code>chr1</code>, 
2) position in that contig e.g. <code>chr1:10e6-20e6</code>; 
3) have a particular contig mapping e.g. <code>chr1,chr2</code>; 
4) interval mapping in both contigs e.g. <code>chr1:10e3-10e6,chr2:0-10e6</code></p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>tomahawk view -i 1kgp3_chr6_1_45.two -I <span class="m">6</span>:10e3-10e6,6:0-10e6 -H <span class="p">|</span> head -n <span class="m">6</span>
</pre></div>
</td></tr></table>

<p>Viewing all data</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>tomahawk view -i 1kgp3_chr6_1_45.two -H <span class="p">|</span> head -n <span class="m">6</span>
</pre></div>
</td></tr></table>

<table>
<thead>
<tr>
<th>FLAG</th>
<th>CHROM_A</th>
<th>POS_A</th>
<th>CHROM_B</th>
<th>POS_B</th>
<th>REF_REF</th>
<th>REF_ALT</th>
<th>ALT_REF</th>
<th>ALT_ALT</th>
<th>D</th>
<th>DPrime</th>
<th>R</th>
<th>R2</th>
<th>P</th>
<th>ChiSqModel</th>
<th>ChiSqTable</th>
</tr>
</thead>
<tbody>
<tr>
<td>2059</td>
<td>6</td>
<td>89572</td>
<td>6</td>
<td>214654</td>
<td>4999</td>
<td>6</td>
<td>0</td>
<td>3</td>
<td>0.000597965</td>
<td>1</td>
<td>0.577004</td>
<td>0.332934</td>
<td>4.01511e-09</td>
<td>1667.33</td>
<td>0</td>
</tr>
<tr>
<td>2059</td>
<td>6</td>
<td>89573</td>
<td>6</td>
<td>214654</td>
<td>4999</td>
<td>6</td>
<td>0</td>
<td>3</td>
<td>0.000597965</td>
<td>1</td>
<td>0.577004</td>
<td>0.332934</td>
<td>4.01511e-09</td>
<td>1667.33</td>
<td>0</td>
</tr>
<tr>
<td>2059</td>
<td>6</td>
<td>122855</td>
<td>6</td>
<td>214654</td>
<td>4988</td>
<td>17</td>
<td>0</td>
<td>3</td>
<td>0.000596649</td>
<td>1</td>
<td>0.38664</td>
<td>0.149491</td>
<td>5.44908e-08</td>
<td>748.648</td>
<td>0</td>
</tr>
<tr>
<td>3</td>
<td>6</td>
<td>143500</td>
<td>6</td>
<td>212570</td>
<td>4421</td>
<td>387</td>
<td>82</td>
<td>118</td>
<td>0.0195352</td>
<td>0.54402</td>
<td>0.331324</td>
<td>0.109775</td>
<td>1.53587e-69</td>
<td>549.755</td>
<td>0</td>
</tr>
<tr>
<td>3</td>
<td>6</td>
<td>143500</td>
<td>6</td>
<td>213499</td>
<td>4419</td>
<td>387</td>
<td>84</td>
<td>118</td>
<td>0.0194949</td>
<td>0.537523</td>
<td>0.329068</td>
<td>0.108286</td>
<td>7.57426e-69</td>
<td>542.296</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>Slicing a range</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>tomahawk view -i 1kgp3_chr6_1_45.two -H -I <span class="m">6</span>:5e6-6e6 <span class="p">|</span> head -n <span class="m">6</span>
</pre></div>
</td></tr></table>

<table>
<thead>
<tr>
<th>FLAG</th>
<th>CHROM_A</th>
<th>POS_A</th>
<th>CHROM_B</th>
<th>POS_B</th>
<th>REF_REF</th>
<th>REF_ALT</th>
<th>ALT_REF</th>
<th>ALT_ALT</th>
<th>D</th>
<th>DPrime</th>
<th>R</th>
<th>R2</th>
<th>P</th>
<th>ChiSqModel</th>
<th>ChiSqTable</th>
</tr>
</thead>
<tbody>
<tr>
<td>2063</td>
<td>6</td>
<td>5022893</td>
<td>6</td>
<td>73938</td>
<td>5000</td>
<td>7</td>
<td>0</td>
<td>1</td>
<td>0.000199362</td>
<td>1</td>
<td>0.353306</td>
<td>0.124825</td>
<td>0.00159744</td>
<td>625.125</td>
<td>0</td>
</tr>
<tr>
<td>2063</td>
<td>6</td>
<td>5020564</td>
<td>6</td>
<td>89339</td>
<td>5000</td>
<td>7</td>
<td>0</td>
<td>1</td>
<td>0.000199362</td>
<td>1</td>
<td>0.353306</td>
<td>0.124825</td>
<td>0.00159744</td>
<td>625.125</td>
<td>0</td>
</tr>
<tr>
<td>7</td>
<td>6</td>
<td>5018459</td>
<td>6</td>
<td>156100</td>
<td>1150</td>
<td>1283</td>
<td>388</td>
<td>2187</td>
<td>0.0804323</td>
<td>0.509361</td>
<td>0.348864</td>
<td>0.121706</td>
<td>1.17649e-138</td>
<td>609.504</td>
<td>0</td>
</tr>
<tr>
<td>7</td>
<td>6</td>
<td>5018691</td>
<td>6</td>
<td>156100</td>
<td>861</td>
<td>811</td>
<td>677</td>
<td>2659</td>
<td>0.0693918</td>
<td>0.339199</td>
<td>0.31898</td>
<td>0.101748</td>
<td>1.10017e-109</td>
<td>509.554</td>
<td>0</td>
</tr>
<tr>
<td>7</td>
<td>6</td>
<td>5018910</td>
<td>6</td>
<td>156100</td>
<td>861</td>
<td>811</td>
<td>677</td>
<td>2659</td>
<td>0.0693918</td>
<td>0.339199</td>
<td>0.31898</td>
<td>0.101748</td>
<td>1.10017e-109</td>
<td>509.554</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>Slicing matches in B string</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>tomahawk view -i 1kgp3_chr6_1_45.two -H -I <span class="m">6</span>:5e6-6e6,6:5e6-6e6 <span class="p">|</span> head -n <span class="m">6</span>
</pre></div>
</td></tr></table>

<table>
<thead>
<tr>
<th>FLAG</th>
<th>CHROM_A</th>
<th>POS_A</th>
<th>CHROM_B</th>
<th>POS_B</th>
<th>REF_REF</th>
<th>REF_ALT</th>
<th>ALT_REF</th>
<th>ALT_ALT</th>
<th>D</th>
<th>DPrime</th>
<th>R</th>
<th>R2</th>
<th>P</th>
<th>ChiSqModel</th>
<th>ChiSqTable</th>
</tr>
</thead>
<tbody>
<tr>
<td>3</td>
<td>6</td>
<td>5000012</td>
<td>6</td>
<td>5073477</td>
<td>4988</td>
<td>9</td>
<td>5</td>
<td>6</td>
<td>0.0011915</td>
<td>0.544089</td>
<td>0.465743</td>
<td>0.216917</td>
<td>1.05037e-13</td>
<td>1086.32</td>
<td>0</td>
</tr>
<tr>
<td>1027</td>
<td>6</td>
<td>5000160</td>
<td>6</td>
<td>5072168</td>
<td>5000</td>
<td>2</td>
<td>4</td>
<td>2</td>
<td>0.000398404</td>
<td>0.4994</td>
<td>0.407677</td>
<td>0.166201</td>
<td>7.1708e-06</td>
<td>832.333</td>
<td>0</td>
</tr>
<tr>
<td>1027</td>
<td>6</td>
<td>5000160</td>
<td>6</td>
<td>5078340</td>
<td>5000</td>
<td>2</td>
<td>4</td>
<td>2</td>
<td>0.000398404</td>
<td>0.4994</td>
<td>0.407677</td>
<td>0.166201</td>
<td>7.1708e-06</td>
<td>832.333</td>
<td>0</td>
</tr>
<tr>
<td>3</td>
<td>6</td>
<td>5000482</td>
<td>6</td>
<td>5079621</td>
<td>4993</td>
<td>6</td>
<td>2</td>
<td>7</td>
<td>0.0013931</td>
<td>0.777199</td>
<td>0.64641</td>
<td>0.417846</td>
<td>3.9492e-18</td>
<td>2092.57</td>
<td>0</td>
</tr>
<tr>
<td>3</td>
<td>6</td>
<td>5000482</td>
<td>6</td>
<td>5082227</td>
<td>4994</td>
<td>6</td>
<td>1</td>
<td>7</td>
<td>0.00139362</td>
<td>0.874675</td>
<td>0.685808</td>
<td>0.470333</td>
<td>8.78523e-19</td>
<td>2355.43</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>As alluded to in the <a href="#sorting-output-files">sorting section</a>, sorted files have 
generally much faster query times. We can demonstrate this difference by using the
sorted and unsorted data from the sliding window example above.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>$ <span class="nb">time</span> tomahawk view -i 1kgp3_chr6_4mb.two -H -I <span class="m">6</span>:1e6-5e6,6:1e6-5e6 &gt; /dev/null

<span class="hll">real    3m1.029s
</span>user    2m41.929s
sys     0m5.682s
</pre></div>
</td></tr></table>

<p>Sorted</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>$ <span class="nb">time</span> tomahawk view -i 1kgp3_chr6_4mb_sorted.two -H -I <span class="m">6</span>:1e6-5e6,6:1e6-5e6 &gt; /dev/null

<span class="hll">real    0m38.167s
</span>user    0m37.579s
sys     0m0.192s
</pre></div>
</td></tr></table>

<p>Viewing <code>ld</code> data from the binary <code>two</code> file format and filtering out lines with a
Fisher's exact test P-value &lt; 1e-4, minor haplotype frequency &lt; 5 and have
both markers on the same contig (bit <code>2</code>)</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>tomahawk view -i file.two -P 1e-4 -a <span class="m">5</span> -f <span class="m">2</span>
</pre></div>
</td></tr></table>

<p>Example output</p>
<table>
<thead>
<tr>
<th>FLAG</th>
<th>CHROM_A</th>
<th>POS_A</th>
<th>CHROM_B</th>
<th>POS_B</th>
<th>REF_REF</th>
<th>REF_ALT</th>
<th>ALT_REF</th>
<th>ALT_ALT</th>
<th>D</th>
<th>Dprime</th>
<th>R</th>
<th>R2</th>
<th>P</th>
<th>ChiSqModel</th>
<th>ChiSqTable</th>
</tr>
</thead>
<tbody>
<tr>
<td>15</td>
<td>20</td>
<td>1314874</td>
<td>20</td>
<td>2000219</td>
<td>5002</td>
<td>5</td>
<td>0</td>
<td>1</td>
<td>0.00019944127</td>
<td>1</td>
<td>0.4080444</td>
<td>0.16650023</td>
<td>0.0011980831</td>
<td>0</td>
<td>833.83313</td>
</tr>
<tr>
<td>15</td>
<td>20</td>
<td>1315271</td>
<td>20</td>
<td>1992301</td>
<td>5005</td>
<td>2</td>
<td>0</td>
<td>1</td>
<td>0.00019956089</td>
<td>1</td>
<td>0.57723492</td>
<td>0.33320019</td>
<td>0.00059904153</td>
<td>0</td>
<td>1668.6665</td>
</tr>
<tr>
<td>15</td>
<td>20</td>
<td>1315527</td>
<td>20</td>
<td>1991024</td>
<td>5004</td>
<td>0</td>
<td>3</td>
<td>1</td>
<td>0.00019952102</td>
<td>1</td>
<td>0.49985018</td>
<td>0.2498502</td>
<td>0.00079872204</td>
<td>0</td>
<td>1251.2498</td>
</tr>
<tr>
<td>15</td>
<td>20</td>
<td>1315763</td>
<td>20</td>
<td>1982489</td>
<td>5006</td>
<td>0</td>
<td>1</td>
<td>1</td>
<td>0.00019960076</td>
<td>1</td>
<td>0.70703614</td>
<td>0.49990013</td>
<td>0.00039936102</td>
<td>0</td>
<td>2503.4999</td>
</tr>
<tr>
<td>15</td>
<td>20</td>
<td>1315807</td>
<td>20</td>
<td>1982446</td>
<td>5004</td>
<td>3</td>
<td>0</td>
<td>1</td>
<td>0.00019952102</td>
<td>1</td>
<td>0.49985018</td>
<td>0.2498502</td>
<td>0.00079872204</td>
<td>0</td>
<td>1251.2498</td>
</tr>
</tbody>
</table>
<p>Example unphased output. Notice that estimated haplotype counts are now floating values and the bit-flag 1 is not set.</p>
<table>
<thead>
<tr>
<th>FLAG</th>
<th>CHROM_A</th>
<th>POS_A</th>
<th>CHROM_B</th>
<th>POS_B</th>
<th>REF_REF</th>
<th>REF_ALT</th>
<th>ALT_REF</th>
<th>ALT_ALT</th>
<th>D</th>
<th>Dprime</th>
<th>R</th>
<th>R2</th>
<th>P</th>
<th>ChiSqModel</th>
<th>ChiSqTable</th>
</tr>
</thead>
<tbody>
<tr>
<td>46</td>
<td>20</td>
<td>1882564</td>
<td>20</td>
<td>1306588</td>
<td>5003.9996</td>
<td>2.0003999</td>
<td>1.0003999</td>
<td>0.99960008</td>
<td>0.00019936141</td>
<td>0.49950022</td>
<td>0.40779945</td>
<td>0.1663004</td>
<td>0.0011978438</td>
<td>0.0011996795</td>
<td>832.83241</td>
</tr>
<tr>
<td>46</td>
<td>20</td>
<td>1895185</td>
<td>20</td>
<td>1306588</td>
<td>5004.9998</td>
<td>1.0001999</td>
<td>1.0001999</td>
<td>0.99980011</td>
<td>0.0001994811</td>
<td>0.49970025</td>
<td>0.49970022</td>
<td>0.24970032</td>
<td>0.00079864228</td>
<td>0.00069960005</td>
<td>1250.4992</td>
</tr>
<tr>
<td>46</td>
<td>20</td>
<td>1306588</td>
<td>20</td>
<td>1901581</td>
<td>5003.9996</td>
<td>2.0003999</td>
<td>1.0003999</td>
<td>0.99960008</td>
<td>0.00019936141</td>
<td>0.49950022</td>
<td>0.40779945</td>
<td>0.1663004</td>
<td>0.0011978438</td>
<td>0.0011996795</td>
<td>832.83241</td>
</tr>
<tr>
<td>46</td>
<td>20</td>
<td>1901581</td>
<td>20</td>
<td>1306588</td>
<td>5003.9996</td>
<td>2.0003999</td>
<td>1.0003999</td>
<td>0.99960008</td>
<td>0.00019936141</td>
<td>0.49950022</td>
<td>0.40779945</td>
<td>0.1663004</td>
<td>0.0011978438</td>
<td>0.0011996795</td>
<td>832.83241</td>
</tr>
<tr>
<td>46</td>
<td>20</td>
<td>1306649</td>
<td>20</td>
<td>1885268</td>
<td>5006</td>
<td>1</td>
<td>1.2656777e-08</td>
<td>0.99999999</td>
<td>0.00019960076</td>
<td>1</td>
<td>0.70703614</td>
<td>0.49990013</td>
<td>0.00039936102</td>
<td>0.00039969285</td>
<td>2503.4999</td>
</tr>
</tbody>
</table>
<p>Example output for forced phased math in fast mode. Note that only the <code>REF_REF</code> count is available, the fast math bit-flag is set, and all P and Chi-squared CV values are 0.</p>
<table>
<thead>
<tr>
<th>FLAG</th>
<th>CHROM_A</th>
<th>POS_A</th>
<th>CHROM_B</th>
<th>POS_B</th>
<th>REF_REF</th>
<th>REF_ALT</th>
<th>ALT_REF</th>
<th>ALT_ALT</th>
<th>D</th>
<th>Dprime</th>
<th>R</th>
<th>R2</th>
<th>P</th>
<th>ChiSqModel</th>
<th>ChiSqTable</th>
</tr>
</thead>
<tbody>
<tr>
<td>67</td>
<td>20</td>
<td>1345922</td>
<td>20</td>
<td>1363176</td>
<td>4928</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0.011788686</td>
<td>0.98334378</td>
<td>0.86251211</td>
<td>0.74392718</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr>
<td>67</td>
<td>20</td>
<td>1345922</td>
<td>20</td>
<td>1367160</td>
<td>4933</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0.011800847</td>
<td>0.98336071</td>
<td>0.89164203</td>
<td>0.79502547</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr>
<td>67</td>
<td>20</td>
<td>1345958</td>
<td>20</td>
<td>1348347</td>
<td>4944</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0.0092644105</td>
<td>0.97890127</td>
<td>0.85316211</td>
<td>0.72788554</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr>
<td>67</td>
<td>20</td>
<td>1345958</td>
<td>20</td>
<td>1354524</td>
<td>4938</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0.0092493389</td>
<td>0.86871892</td>
<td>0.80354655</td>
<td>0.64568704</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr>
<td>75</td>
<td>20</td>
<td>1345958</td>
<td>20</td>
<td>1356626</td>
<td>4945</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0.0033518653</td>
<td>0.99999994</td>
<td>0.51706308</td>
<td>0.26735422</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
</tbody>
</table>
<h2 id="aggregating-and-visualizing-datasets">Aggregating and visualizing datasets<a class="headerlink" href="#aggregating-and-visualizing-datasets" title="Permanent link">&para;</a></h2>
<p>Tomahawk generally output many millions to many hundreds of millions to billions
of output linkage disequilibrium (LD) associations generated from many millions
of input SNVs. It is technically very challenging to visualize such large
datasets. Read more about <a href="../aggregation/">aggregation</a> in Tomahawk.</p>
<table>
<thead>
<tr>
<th>Aggregation</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>R</code></td>
<td>Pearson correlation coefficient</td>
</tr>
<tr>
<td><code>R2</code></td>
<td>Squared pearson correlation coefficient</td>
</tr>
<tr>
<td><code>D</code></td>
<td>Coefficient of linkage disequilibrium</td>
</tr>
<tr>
<td><code>Dprime</code></td>
<td>Scaled coefficient of linkage disequilibrium</td>
</tr>
<tr>
<td><code>Dp</code></td>
<td>Alias for <code>dprime</code></td>
</tr>
<tr>
<td><code>P</code></td>
<td>Fisher's exact test P-value of the 2x2 haplotype contigency table</td>
</tr>
<tr>
<td><code>Hets</code></td>
<td>Number of (0,1) or (1,0) associations</td>
</tr>
<tr>
<td><code>Alts</code></td>
<td>Number of (1,1) associations</td>
</tr>
<tr>
<td><code>Het</code></td>
<td>Alias for <code>hets</code></td>
</tr>
<tr>
<td><code>Alt</code></td>
<td>Alias for <code>alts</code></td>
</tr>
</tbody>
</table>
<table>
<thead>
<tr>
<th>Reduction</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>Mean</code></td>
<td>Mean number of aggregate</td>
</tr>
<tr>
<td><code>Max</code></td>
<td>Largest number in aggregate bin</td>
</tr>
<tr>
<td><code>Min</code></td>
<td>Smallest number in aggregate bin</td>
</tr>
<tr>
<td><code>Count</code></td>
<td>Total number of records in a bin</td>
</tr>
<tr>
<td><code>N</code></td>
<td>Alias for <code>count</code></td>
</tr>
<tr>
<td><code>Total</code></td>
<td>Sum total of aggregated number in a bin</td>
</tr>
</tbody>
</table>
<p>In this example, we will aggregate the genome-wide data generated from the sliding example
above by aggregating on <code>R2</code> values and reducing into <code>count</code> in a (4000,4000)-bin space. If
a bin have less than 50 observations (<code>-c</code>) we will drop all the value and report 0 for that bin.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>twk aggregate -i 1kgp3_chr6_4mb_sorted.two -x <span class="m">4000</span> -y <span class="m">4000</span> -f r2 -r count -c <span class="m">50</span> -t <span class="m">4</span> -O b -o 1kgp3_chr6_4mb_aggregate.twa
</pre></div>
</td></tr></table>

<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>Program:   tomahawk-7f8eef9b-dirty (Tools for computing, querying and storing LD data)
Libraries: tomahawk-0.7.0; ZSTD-1.3.8; htslib 1.9
Contact: Marcus D. R. Klarqvist &lt;mk819@cam.ac.uk&gt;
Documentation: https://github.com/mklarqvist/tomahawk
License: MIT
----------
[2019-01-25 14:55:43,299][LOG] Calling aggregate...
[2019-01-25 14:55:43,299][LOG] Performing 2-pass over data...
[2019-01-25 14:55:43,299][LOG] ===== First pass (peeking at landscape) =====
[2019-01-25 14:55:43,299][LOG] Blocks: 47,352
[2019-01-25 14:55:43,299][LOG] Uncompressed size: 50.192950 Gb
[2019-01-25 14:55:43,299][LOG][THREAD] Data/thread: 12.548238 Gb
[2019-01-25 14:55:43,299][PROGRESS] Time elapsed       Variants  Progress   Est. Time left
[2019-01-25 14:56:13,299][PROGRESS]      30,000s    344,930,000   72.8446%  11s
[2019-01-25 14:56:27,979][PROGRESS] 44,679s 473,514,826 (10,597,935 variants/s)
[2019-01-25 14:56:27,979][PROGRESS] Finished!
[2019-01-25 14:56:27,979][LOG] ===== Second pass (building matrix) =====
[2019-01-25 14:56:27,979][LOG] Aggregating 473,514,826 records...
[2019-01-25 14:56:27,979][LOG][THREAD] Allocating: 2.560000 Gb for matrices...
[2019-01-25 14:56:27,979][PROGRESS] Time elapsed       Variants  Progress   Est. Time left
[2019-01-25 14:56:57,979][PROGRESS]      30,000s    348,560,000   73.6112%  10s
[2019-01-25 14:57:12,499][PROGRESS] 44,520s 473,514,826 (10,635,926 variants/s)
[2019-01-25 14:57:12,499][PROGRESS] Finished!
[2019-01-25 14:57:13,713][LOG] Aggregated 473,514,826 records in 16,000,000 bins.
[2019-01-25 14:57:13,713][LOG] Finished.
</pre></div>
</td></tr></table>

<p>You can now use the output binary format <code>.twa</code> in your downstream analysis.
Alternatively, it is possible to output a human-readable (x,y)-matrix by setting
the <code>-O</code> parameter to <code>u</code>. This tab-delimited matrix can now be loaded in any
programming language and used as input for graphical visualizations or for analysis.
It is easiest to use these files directly in <a href="../r-tutorial/"><code>rtomahawk</code></a>, the R-bindings for <code>tomahawk</code>.</p>
                
                  
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href="../job-balancing/" title="Job balancing" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Job balancing
              </span>
            </div>
          </a>
        
        
          <a href="../r-tutorial/" title="R" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                R
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            Copyright (C) 2016-present Genome Research Ltd., University of Cambridge, and Marcus D. R. Klarqvist
          </div>
        
        powered by
        <a href="https://www.mkdocs.org">MkDocs</a>
        and
        <a href="https://squidfunk.github.io/mkdocs-material/">
          Material for MkDocs</a>
      </div>
      
        
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../assets/javascripts/application.b3a2063f.js"></script>
      
      <script>app.initialize({version:"1.0.4",url:{base:".."}})</script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML"></script>
      
    
    
      
    
  </body>
</html>