\documentclass[10pt]{article}
\usepackage[margin=1in]{geometry}
\usepackage{color}
\definecolor{gray}{rgb}{0.7,0.7,0.7}
\usepackage{framed}
\usepackage{enumitem}
\usepackage{longtable}
\usepackage{makecell}
\usepackage{amstext}
\usepackage{float}
\usepackage[pdfborder={0 0 0},hyperfootnotes=false]{hyperref}

\begin{document}

\section{The TWK Format Specification}

\subsection{TGZF specification}
TGZF is a variation of BGZF used in the BAM and BCF formats (https://samtools.github.io/hts-specs/) that in turn are slightly modified gzip blocks concatenated together back-to-back to enable random access lookups. In short, TGZF lifts the size restriction in BGZF that uncompressed data cannot exceed $2^{16}$  bytes and instead asserts that the compressed TGZF data described the BSIZE field is smaller than $2^{32}$ bytes. When a block of uncompressed data is dispatched to the compressor and sent to disk is contextually determined and is available as a modifiable variable in many cases.

Notably, as the TGZF/BGZF format naturally extend the ZLIB library (http://zlib.net/), all multi-byte values are little endian (as required by the gzip specification). Endianness is not asserted in the uncompressed data.

\begin{table}[H]
{\small
\begin{tabular}{|l|l|l|l|l|r|}
  \cline{1-6}
  \multicolumn{3}{|c|}{\bf Field} & \multicolumn{1}{c|}{\bf Description} & \multicolumn{1}{c|}{\bf Type} & \multicolumn{1}{c|}{\bf Value} \\\cline{1-6}
  \multicolumn{6}{|c|}{\textcolor{gray}{\it List of compression blocks (until the end of the file)}} \\\cline{2-6}
  & \multicolumn{2}{l|}{\sf ID1} & gzip IDentifier1 & {\tt uint8\_t} & 31 \\\cline{2-6}
  & \multicolumn{2}{l|}{\sf ID2} & gzip IDentifier2 & {\tt uint8\_t} & 139 \\\cline{2-6}
  & \multicolumn{2}{l|}{\sf CM} & gzip Compression Method & {\tt uint8\_t} & 8 \\\cline{2-6}
  & \multicolumn{2}{l|}{\sf FLG} & gzip FLaGs & {\tt uint8\_t} & 4 \\\cline{2-6}
  & \multicolumn{2}{l|}{\sf MTIME} & gzip Modification TIME & {\tt uint32\_t} & \\\cline{2-6}
  & \multicolumn{2}{l|}{\sf XFL} & gzip eXtra FLags & {\tt uint8\_t} & 0 \\\cline{2-6}
  & \multicolumn{2}{l|}{\sf OS} & gzip Operating System & {\tt uint8\_t} & 255 \\\cline{2-6}
  & \multicolumn{2}{l|}{\sf XLEN} & gzip eXtra LENgth & {\tt uint16\_t} & 8 \\\cline{2-6}
  & \multicolumn{5}{c|}{\textcolor{gray}{\it Extra subfield(s) (total size=XLEN)}} \\\cline{3-6}
  & & \multicolumn{4}{c|}{\textcolor{gray}{\it Additional RFC1952 extra subfields if present}} \\\cline{3-6}
  & & {\sf SI1} & Subfield Identifier1 & {\tt uint8\_t} & 84 \\\cline{3-6}
  & & {\sf SI2} & Subfield Identifier2 & {\tt uint8\_t} & 90 \\\cline{3-6}
  & & {\sf SLEN} & Subfield LENgth & {\tt uint16\_t} & 2 \\\cline{3-6}
  & & {\sf BSIZE} & total Block SIZE & {\tt uint32\_t} & \\\cline{3-6}
  & & \multicolumn{4}{c|}{\textcolor{gray}{\it Additional RFC1952 extra subfields if present}} \\\cline{2-6}
  & \multicolumn{2}{l|}{\sf CDATA} & Compressed DATA by {\sf zlib::deflate()} & {\tt uint8\_t[{\sf BSIZE-XLEN-19}]} & \\\cline{2-6}
  & \multicolumn{2}{l|}{\sf CRC32} & CRC-32 & {\tt uint32\_t} & \\\cline{2-6}
  & \multicolumn{2}{l|}{\sf ISIZE} & Input SIZE (length of uncompressed data) & {\tt uint32\_t} & \\
  \cline{1-6}
\end{tabular}}
\end{table}

\subsection{Template type data bounds}
The Tomahawk data structure described below uses different integer types based on the number of samples in the imported file. This Template field is reinterpreted from the byte buffer as follows:
\begin{table}[H]
  {\small
\begin{tabular}{|l|l|}
  \cline{1-2}
  {\bf Field} & {\bf Description} \\ \cline{1-2}
  {\sf uint8\_t} & {\sf 0 $<$ n\_samples $<$ 16 ($2^4$)} \\ \cline{1-2}
  {\sf uint16\_t} & {\sf 16 $<$ n\_samples $<$ 4096 ($2^{12}$)} \\ \cline{1-2}
  {\sf uint32\_t} & {\sf 4096 $<$ n\_samples $<$ 268435456 ($2^{28}$)} \\ \cline{1-2}
  {\sf uint64\_t} & {\sf 268435456 $<$ n\_samples $<$ 1152921504606846976 ($2^{60}$)} \\ \cline{1-2}
  \end{tabular}}
\end{table}
Using this approach, Tomahawk run-length encodes genotypes with the lowest 4 bits being the alleles and the remainder $\text{sizeof(Template)} * 8 - 4$ bits as the run-length.

\subsection{TWK organization}
A TWK file comprises of a binary (non-compressed) header followed by a series of TGZF-compressed binary blocks of TWK records and ended with a binary EOF marker.

The TWK format asserts that:
\begin{itemize}
\item Contig information is specified in the header: this data must minimally include a contig identifier (such as a unique string name in VCF or unique integer in BCF) and contig length in base pairs
\item Entries must be bi-allelic SNVs
\item Entries in the imported VCF / BCF file are sorted by contig identifier followed by genomic coordinates (base position of SNV)
\item There is $>$ 1 sample in the file
\item If importing from the BCF format: is version 2.2 or later
\item If importing from the VCF format: is version 4 or later
\end{itemize}

The fields INFO and FORMAT are dropped from imported VCF/BCF files as they are not used in the current implementation

\subsection{TWI: Random access}
Searching to the beginning of a specified TGZF block is aided by TWI entries specifying virtual offsets into the TWK file. Importantly, unlike BAM and BCF, none of TWK/TWI/TWO/TOI permits data to be split over multiple blocks. This intentional restriction guarantees that the uncompressed data is completely disjoint and this assertion renders parallel computing exceedingly more tractable.

Unlike BCI/BAI, TWI/TOI stores the virtual file offset to every single TGZF block. This permits complete random access to any part of a TWK/TWO file. This makes TWI/TOI indices larger than BCI/BAI indices but still remain very small. For example, the TWI file for HRC.v1-1 is $<$ 2 MB.

\begin{table}[H]
{\small
\begin{tabular}{|l|l|l|l|r|}
  \cline{1-5}
  \multicolumn{2}{|l|}{\bf Field} & \multicolumn{1}{l|}{\bf Description} & \multicolumn{1}{l|}{\bf Type} & \multicolumn{1}{r|}{\bf Value} \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf MAGIC} & Start of file identifier string & {\tt char[10]} & {\tt TOTEMPOLE\char92 1} \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf version} & Tomahawk major version & {\tt float} & \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf samples} & Number of samples & {\tt uint64\_t} & \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf controller} & Currently unused & {\tt uint8\_t} & 0 \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf n\_blocks} & Number of TGZF blocks in Twk file & {\tt uint32\_t} & \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf n\_largest} & Size in bytes of largest uncompressed TGZF block & {\tt uint32\_t} & \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf header\_offset} & Relative disk offset until this position (start of data) & {\tt uint32\_t} & \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf n\_contigs} & Number of contigs in header & {\tt uint32\_t} & \\\cline{1-5}
  \multicolumn{5}{|c|}{\textcolor{gray}{\it List of contig data (n\_contigs)}} \\\cline{2-5}
  & {\sf bp\_contig} & Length of contig in bases & {\tt uint32\_t} & \\\cline{2-5}
  & {\sf l\_contig} & Length of contig name & {\tt uint32\_t} & \\\cline{2-5}
  & {\sf contig\_name} & Contig name & {\tt char[l\_contig]} & \\\cline{2-5}
  \multicolumn{5}{|c|}{\textcolor{gray}{\it List of sample identifiers (n\_samples)}} \\\cline{2-5}
  & {\sf l\_name} & Length of sample name & {\tt uint32\_t} & \\\cline{2-5}
  & {\sf sample\_name} & Sample name & {\tt char[l\_name]} & \\\cline{2-5}
  \multicolumn{2}{|l|}{\sf TGZF\_block} & \begin{tabular}{@{}l@{}}Compressed DATA by {\sf zlib::deflate()}. DATA keeping \\ VCF header and any changes made to the file\end{tabular} &  & \\\cline{1-5}
  \multicolumn{5}{|c|}{\textcolor{gray}{\it Totempole entries until end-of-file}} \\\cline{2-5}
  & {\sf byte\_offset} & Virtual file offset to start to TGZF block & {\tt uint64\_t} & \\\cline{2-5}
  & {\sf contigID} & All variants belong to this contig identifier & {\tt int32\_t} & \\\cline{2-5}
    & {\sf min\_position} & Smallest variant position & {\tt uint32\_t} & \\\cline{2-5}
  & {\sf max\_position} & Largest variant position & {\tt uint32\_t} & \\\cline{2-5}
    & {\sf n\_variants} & Number of variants & {\tt uint16\_t} & \\\cline{2-5}
  & {\sf uncompressed\_size} & Uncompressed size of data & {\tt uint32\_t} & \\\cline{2-5}
  \multicolumn{2}{|l|}{\sf EOF\_string} & End-of-file marker & {\tt char*} & \\\cline{1-5}
\end{tabular}}
\end{table}

\subsection{TWK format}
\begin{table}[H]
{\small
\begin{tabular}{|l|l|l|l|r|}
  \cline{1-5}
  \multicolumn{2}{|l|}{\bf Field} & \multicolumn{1}{l|}{\bf Description} & \multicolumn{1}{l|}{\bf Type} & \multicolumn{1}{r|}{\bf Value} \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf MAGIC} & Start of file identifier string & {\tt char[9]} & {\tt TOMAHAWK\char92 1} \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf version} & Tomahawk major version & {\tt float} & \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf samples} & Number of samples & {\tt uint64\_t} & \\\cline{1-5}

  \multicolumn{5}{|c|}{\textcolor{gray}{\it TGZF blocks until end-of-file}} \\\cline{2-5}
  & \multicolumn{4}{|l|}{\textcolor{gray}{\it For n\_variants (described in TWI)}} \\\cline{2-5}
  & {\sf pos\_plus} & \begin{tabular}{@{}l@{}}{\tt{\sf pos}\char60\char60 30\char124\tt{\sf phased}\char60\char60 1\char124{\sf missing}}; Genomic coordinate; \\ flag if all data is phased; flag if any data is missing\end{tabular} & {\tt uint32\_t} & \\\cline{2-5}
  & {\sf ref\_alt} & \tt{\sf REF}\char60\char60 4\char124\tt{\sf ALT} & {\tt uint8\_t} & \\\cline{2-5}
  & {\sf MAF} & Minor allele frequency & {\tt float} & \\\cline{2-5}
  & {\sf HWE\_P} & Hardy-Weinberg P-value (Fisher's exact test) & {\tt float} & \\\cline{2-5}
  & {\sf n\_runs} & Number of runs for this variant & {\tt Template} & \\\cline{2-5}
  & \multicolumn{4}{|l|}{\textcolor{gray}{\it Until end of TGZF block}} \\\cline{2-5}
  & {\sf RLE} & Run-length encoded data & {\tt Template*} & \\\cline{2-5}
  \multicolumn{2}{|l|}{\sf EOF\_string} & End-of-file marker & {\tt char*} & \\\cline{1-5}
\end{tabular}}
\end{table}

\subsection{TWO format}
TWO entry
\begin{table}[H]
{\small
\begin{tabular}{|l|l|l|l|r|}
  \cline{1-5}
  \multicolumn{2}{|l|}{\bf Field} & \multicolumn{1}{l|}{\bf Description} & \multicolumn{1}{l|}{\bf Type} & \multicolumn{1}{r|}{\bf Value} \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf MAGIC} & Start of file identifier string & {\tt char[17]} & {\tt TOMAHAWK\char`\~OUTPUT\char92 1} \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf version} & Tomahawk major version & {\tt float} & \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf samples} & Number of samples & {\tt uint64\_t} & \\\cline{1-5}
   \multicolumn{2}{|l|}{\sf n\_contigs} & Number of contigs in header & {\tt uint32\_t} & \\\cline{1-5}

  \multicolumn{5}{|c|}{\textcolor{gray}{\it List of contig data (n\_contigs)}} \\\cline{2-5}
  & {\sf bp\_contig} & Length of contig in bases & {\tt uint32\_t} & \\\cline{2-5}
  & {\sf l\_contig} & Length of contig name & {\tt uint32\_t} & \\\cline{2-5}
  & {\sf contig\_name} & Contig name & {\tt char[l\_contig]} & \\\cline{2-5}

   \multicolumn{2}{|l|}{\sf TGZF\_block} & \begin{tabular}{@{}l@{}}Compressed DATA by {\sf zlib::deflate()}. DATA keeping \\ VCF header and any changes made to the file\end{tabular} &  & \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf EOF\_string} & End-of-file marker & {\tt char*} & \\\cline{1-5}
   \multicolumn{5}{|c|}{\textcolor{gray}{\it TGZF blocks of TWO entries until EOF}} \\\cline{1-5}
\end{tabular}}
\end{table}

TWO entries
\begin{table}[H]
{\small
\begin{tabular}{|l|l|l|l|r|}
  \cline{1-5}
  \multicolumn{2}{|l|}{\bf Field} & \multicolumn{1}{l|}{\bf Description} & \multicolumn{1}{l|}{\bf Type} & \multicolumn{1}{r|}{\bf Value} \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf FLAG} & Bit-wise flags & {\tt uint16\_t} & \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf AcontigID} & Variant A contig map identifier, 0 $\leq$ AcontigID $<$ n\_ref & {\tt uint32\_t} & \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf Aposition} & \tt{\sf pos}\char60\char60 30\char124\tt{\sf phased}\char60\char60 1\char124{\sf missing} & {\tt uint32\_t} & \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf BcontigID} & Variant B contig map identifier, 0 $\leq$ BcontigID $<$ n\_ref & {\tt uint32\_t} & \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf Bposition} & \tt{\sf pos}\char60\char60 30\char124\tt{\sf phased}\char60\char60 1\char124{\sf missing} & {\tt uint32\_t} & \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf p1} & Haplotype counts for A1B1 (Ref-Ref). Is estimated if FLAG bit 1 is unset & {\tt float} & \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf p2} & Haplotype counts for A1B2 (Ref-Alt). Is estimated if FLAG bit 1 is unset & {\tt float} & \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf q1} & Haplotype counts for A2B1 (Alt-Ref). Is estimated if FLAG bit 1 is unset & {\tt float} & \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf q2} & Haplotype counts for A2B2 (Alt-Alt). Is estimated if FLAG bit 1 is unset & {\tt float} & \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf D} & Coefficient of linkage disequilibrium (D) & {\tt float} & \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf Dprime} & Normalised D value & {\tt float} & \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf R2} & Tomahawk major version & {\tt float} & \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf P} & Tomahawk major version & {\tt double} & \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf chiSqFisher} & \begin{tabular}{@{}l@{}}Exact Fisher's test or Chi-squared test (see FLAG bit) \\ for the 2x2 haplotype contingency table\end{tabular} & {\tt double} & \\\cline{1-5}
  \multicolumn{2}{|l|}{\sf chiSqModel} & Chi-squared critical value for the 3x3 genotype contingency table & {\tt double} & \\\cline{1-5}
\end{tabular}}
\end{table}

\subsection{TWO FLAG field}
The TWO FLAG
\begin{table}[H]
{\small
\begin{tabular}{|l|l|}
  \cline{1-2}
  \multicolumn{1}{|l|}{\bf Bit} & \multicolumn{1}{l|}{\bf Description} \\ \cline{1-2}
  \multicolumn{1}{|l|}{\sf 1} & Both variant lines were phased OR equations used for phased genotypes was used \\ \cline{1-2}
  \multicolumn{1}{|l|}{\sf 2} & Either variant has missing values \\ \cline{1-2}
  \multicolumn{1}{|l|}{\sf 3} & A field is incomplete (A1B1, A1B2, A2B1, or A2B2 has 0 observations) \\ \cline{1-2}
  \multicolumn{1}{|l|}{\sf 4} & There are multiple possible biological solutions (valid roots in cubic equation) \\ \cline{1-2}
  \multicolumn{1}{|l|}{\sf 5} & Both variants are on the same contig \\ \cline{1-2}
  \multicolumn{1}{|l|}{\sf 6} & There is > 1 million base pairs between the variants \\ \cline{1-2}
  \multicolumn{1}{|l|}{\sf 7} & There is $>$ 1 million base pairs between the variants \\ \cline{1-2}
  \multicolumn{1}{|l|}{\sf 8} & Variant A failed Hardy-Weinberg test (P $<$ 1E-6) \\ \cline{1-2}
  \multicolumn{1}{|l|}{\sf 9} & Variant B failed Hardy-Weinberg test (P $<$ 1E-6) \\ \cline{1-2}
  \multicolumn{1}{|l|}{\sf 10} & Variant A has a low minor allele frequency ($<$ 1\%) \\ \cline{1-2}
  \multicolumn{1}{|l|}{\sf 11} & Variant B has a low minor allele frequency ($<$ 1\%) \\ \cline{1-2}
  \multicolumn{1}{|l|}{\sf 12} & Currently unused \\ \cline{1-2}
  \multicolumn{1}{|l|}{\sf 13} & Currently unused \\ \cline{1-2}
  \multicolumn{1}{|l|}{\sf 14} & Currently unused \\ \cline{1-2}
  \multicolumn{1}{|l|}{\sf 15} & Currently unused \\ \cline{1-2}
  \multicolumn{1}{|l|}{\sf 16} & Currently unused \\ \cline{1-2}
  \end{tabular}}
\end{table}

\end{document}
